{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63cb6258",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO \n",
    "import os,sys\n",
    "import sgRNA_utils.sgRNA_primer_util as su   \n",
    "import module.sequencing_primer as sr\n",
    "import module.from_gbFile_to_seq as fq\n",
    "import module.parser_input_to_df as pf\n",
    "import module.product_and_decorate_editingSeq as p_d_seq\n",
    "import module.order as order\n",
    "import warnings   \n",
    "warnings.filterwarnings('ignore')     \n",
    "import configparser\n",
    "from Bio import SeqIO  \n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import json\n",
    "from sgRNA_utils.sgRNA_primer_config import config \n",
    "\n",
    "from os.path import exists,splitext,dirname,splitext,basename,realpath,abspath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ff10bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70b93070",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "127013fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = {     \n",
    "        \"chopchop_input\": \"./input/only_primer/info_input.csv\",   \n",
    "        \"sgRNA_result_path\": \"\",\n",
    "        \"edit_sequence_design_workdir\":\"./output/only_primer\",\n",
    "        \"ref_genome\":\"./input/GCA_000011325.1_ASM1132v1_genomic.fna\",\n",
    "        \"one_plasmid_file_path\":\"./input/pXMJ19-Cas9A-gRNA-crtYEb-Ts - ori.gb\",   \n",
    "        \"no_ccdb_plasmid\":\"./input/no-ccdb-pXMJ19-Cas9A-gRNA-crtYEb-Ts - ori.gb\",\n",
    "        \"no_sgRNA_plasmid\":\"./input/no-sgRNA-pXMJ19-Cas9A-gRNA-crtYEb-Ts - ori.gb\",\n",
    "        \"scene\":\"only_primer\",  \n",
    "        \"uha_dha_config\": {\n",
    "            \"max_right_arm_seq_length\": 1050,  \n",
    "            \"max_left_arm_seq_length\": 1050,   \n",
    "            \"min_left_arm_seq_length\": 1000,   \n",
    "            \"min_right_arm_seq_length\": 1000     \n",
    "        },\n",
    "\n",
    "        \"plasmid_label\":{\n",
    "            \"ccdb_label\":\"ccdB\",  \n",
    "            \"promoter_terminator_label\":\"gRNA\",\n",
    "            \"n_20_label\":\"N20\"\n",
    "        },\n",
    "\n",
    "        \"primer_json\":{\n",
    "        \n",
    "        },\n",
    "        \"region_label\":\"\",     \n",
    "\n",
    "        \"sgRNA_primer_json\":{\n",
    "           \n",
    "        },\n",
    "        \"ccdb_primer_json\":{\n",
    "                \n",
    "        },   \n",
    "    \n",
    "        \"sgRNA_region_label\":\"\",\n",
    "        \n",
    "        \"ccdb_region_label\":\"\",   \n",
    "        \n",
    "        \"enzyme\":{\n",
    "            \"enzyme_name\":\"BsaI\",\n",
    "            \"gap_sequence\":\"AA\",  \n",
    "            \"protection_sequence\":\"CCA\"   \n",
    "        },  \n",
    "          \n",
    "        \"UHA_ARGS\":{\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,  \n",
    "            \"PRIMER_MAX_TM\": 75,    \n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "        \"SEQ_ALTERED_ARGS\":{\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,\n",
    "            \"PRIMER_MAX_TM\": 75,  \n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "        \"DHA_ARGS\":{\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,\n",
    "            \"PRIMER_MAX_TM\": 75,\n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "         \"UP_SGRNA_ARGS\": {\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,  \n",
    "            \"PRIMER_MAX_TM\": 75,    \n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "        \"DOWN_SGRNA_ARGS\": {\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,  \n",
    "            \"PRIMER_MAX_TM\": 75,    \n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "\n",
    "        \"PLASMID_Q_ARGS\":{\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,  \n",
    "            \"PRIMER_MAX_TM\": 75,    \n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "        \"GENOME_Q_ARGS\":{\n",
    "            \"PRIMER_OPT_TM\": 65,\n",
    "            \"PRIMER_MIN_TM\": 55,     \n",
    "            \"PRIMER_MAX_TM\": 75,    \n",
    "            \"PRIMER_MIN_GC\": 20,\n",
    "            \"PRIMER_MAX_GC\": 80\n",
    "        },\n",
    "        'sgRNA_result':{\n",
    "            \"Cgl0006_1176_G_A_sub\":\"1\",\n",
    "            \"Cgl2342_213_GCA_ins\":\"1\",\n",
    "            \"Cgl1436_1113_CAA_del\":\"1\",\n",
    "            \"Cgl1790_1647_TCC_sub\":\"1\",\n",
    "            \"Cgl1386_327_18to15_sub\":\"1\",\n",
    "            \"Cgl0591_-1_Ppgk_promoter_ins\":\"1\",\n",
    "            \"Cgl0141_cds_del\":\"1\",\n",
    "            \"153019_ecoil_ybeL_ins\":\"1\",\n",
    "            \"Cgl0851_ecoli_pgi_sub\":\"1\"\n",
    "        }      \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "124f6231",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#uha_dha_primer\n",
    "def extract_uha_dha_primer(info_input_df, sgRNA):\n",
    "\n",
    "   \n",
    "    primer_template_for_u_d_ha_df = p_d_seq.create_primer_template(info_input_df,sgRNA)\n",
    "\n",
    "    uha_dha_primer_df = p_d_seq.design_primer(primer_template_for_u_d_ha_df,'Name_Region','primer_template','u_d')\n",
    "    uha_dha_primer_df = uha_dha_primer_df.rename(columns={'Region':'id'})\n",
    "    uha_dha_primer_df = uha_dha_primer_df.join(uha_dha_primer_df.id.str.split(';',expand=True).rename(columns = {0:'Name',1:'Region',2:'Type'})).drop(columns='id')\n",
    "    \n",
    "    return uha_dha_primer_df\n",
    "\n",
    "#uha_dha \n",
    "def extract_uha_dha(info_input_df,uha_dha_primer_df,sgRNA):\n",
    "    #整合进突变信息\n",
    "    info_df = info_input_df[['name','region','seq_altered','type','ref','strand']].rename(columns={'name':'Name','region':'Region'})\n",
    "    info_df.seq_altered.fillna('',inplace=True)\n",
    "    uha_dha_info_primer_df = pd.merge(info_df,uha_dha_primer_df,on=['Name','Region'])\n",
    "\n",
    "    #提取源生同源臂\n",
    "    uha_dha_df = p_d_seq.create_uha_dha_df(uha_dha_primer_df) \n",
    "    #合并突变信息\n",
    "    uha_dha_df = pd.merge(uha_dha_df,info_df,on=['Name','Region'])\n",
    "    uha_dha_sgRNA_df = pd.merge(uha_dha_df,sgRNA,on=['Name','Region'],how='inner')\n",
    "\n",
    "    return uha_dha_info_primer_df, uha_dha_df, uha_dha_sgRNA_df,info_df\n",
    "\n",
    "def one_plasmid_system_design_by_user_primer(primer_json,primer_position_json,gb_path,plasmid_backbone,enzyme_df,enzyme_name):\n",
    "\n",
    "    sgRNA_promoter_terminator_start = 0\n",
    "    #对引物进行排序,确定所有引物\n",
    "    primer_dict_df, primers_sum = p_d_seq.sort_compose_primer(sgRNA_promoter_terminator_start,\n",
    "                                                         primer_json,\n",
    "                                                         primer_position_json,\n",
    "                                                         gb_path,\n",
    "                                                         plasmid_backbone,\n",
    "                                                         plasmid_type='one')   \n",
    "    #给引物加接头\n",
    "    plasmid_primer_joint_df = p_d_seq.add_joint_plasmid_primer(enzyme_df,enzyme_name,\n",
    "                                                                 primer_dict_df,\n",
    "                                                                 int(primers_sum/2),\n",
    "                                                                 primer_type='ccdb')\n",
    "    #修改df格式\n",
    "    plasmid_primer_df = p_d_seq.plasmid_primer(plasmid_primer_joint_df)\n",
    "    #添加索引\n",
    "    plasmid_primer_df.insert(0,'index',plasmid_primer_df.index)\n",
    "    plasmid_primer_df.reset_index(drop=True, inplace=True)\n",
    "        \n",
    "    #给引物加产物\n",
    "    plasmid_primer_df = p_d_seq.add_product_and_size(gb_path, plasmid_primer_df, enzyme_df,enzyme_name,seq=plasmid_backbone)\n",
    "\n",
    "    return  plasmid_primer_df\n",
    "\n",
    "def one_plasmid_system_design_by_user_region(uha_dha_sgRNA_df,sgRNA_region_seq_json,one_plasmid_file_path,sgRNA_plasmid_backbone,enzyme_df,enzyme_name):\n",
    "    \n",
    "    #序列转换成坐标\n",
    "    region_json = su.convert_seq_cor(one_plasmid_file_path, sgRNA_region_seq_json,seq=sgRNA_plasmid_backbone)\n",
    "    #坐标转转换成距离\n",
    "    distance_dict = region_2_distance(len(sgRNA_plasmid_backbone), region_json, 0)   \n",
    "\n",
    "    #根据区域设计引物\n",
    "    primer_result_list = p_d_seq.design_primer_by_region_in_plasmid(0,sgRNA_plasmid_backbone,distance_dict)\n",
    "    plasmid_primer_df = su.result_primer_list_to_df(primer_result_list)\n",
    "\n",
    "    #加接头\n",
    "    plasmid_primer_joint_df = p_d_seq.sgRNA_sgRNAprimer_merge(uha_dha_sgRNA_df,plasmid_primer_df,sgRNA_columns=['Name','Region'])\n",
    "    primers_sum = len(plasmid_primer_df)\n",
    "    plasmid_primer_joint_df = p_d_seq.add_joint_plasmid_primer(enzyme_df, enzyme_name, plasmid_primer_joint_df, primers_sum, primer_type='ccdb')\n",
    "    plasmid_primer_joint_df = plasmid_primer_joint_df.drop(columns='Region').drop_duplicates()\n",
    "    #修改df格式\n",
    "    plasmid_primer_df = p_d_seq.plasmid_primer(plasmid_primer_joint_df)\n",
    "    #添加索引\n",
    "    plasmid_primer_df.insert(0,'index',plasmid_primer_df.index)\n",
    "    plasmid_primer_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    plasmid_primer_df = p_d_seq.add_product_and_size(one_plasmid_file_path, plasmid_primer_df, enzyme_df, enzyme_name=enzyme_name,seq=sgRNA_plasmid_backbone)\n",
    "\n",
    "    return  plasmid_primer_df\n",
    "\n",
    "#one plasmid pcr primer     \n",
    "def one_plasmid_system_pcr_design_primer(gb_path,\n",
    "                                         info_df,\n",
    "                                         uha_dha_sgRNA_df,\n",
    "                                         uha_dha_info_primer_df,\n",
    "                                         uha_dha_primer_df,\n",
    "                                         enzyme_df,\n",
    "                                         enzyme_name,\n",
    "                                         plasmid_primer_desgin_type,\n",
    "                                         seq_json):\n",
    "    \n",
    "    #创建新的质粒   \n",
    "    uha_dha_sgRNA_df, promoter_terminator_up_promoter_seq, promoter_terminator_down_terminator_seq, type_kind  = p_d_seq.create_new_plasmid(gb_path, uha_dha_sgRNA_df.copy(), ccdb_label='ccdB', promoter_terminator_label='gRNA', n_20_label='N20')\n",
    "\n",
    "    #设计sgRNA、ccdb质粒引物\n",
    "    n20up_primer_template = uha_dha_sgRNA_df[['Name','Region','n20_up_template','Target sequence','Rev Target sequence']]\n",
    "    n20up_primer_template['Region'] = n20up_primer_template['Name'] +';'+ n20up_primer_template['Region']\n",
    "    n20up_primer_df = p_d_seq.design_primer(n20up_primer_template,'Region','n20_up_template','sgRNA')\n",
    "    n20up_primer_df = pd.merge(n20up_primer_template[['Region','Target sequence','Rev Target sequence']],n20up_primer_df,on=['Region'],how='inner')\n",
    "\n",
    "    #ccdb、sgrna质粒引物加接头\n",
    "    n20up_primer_df = p_d_seq.add_joint_sgRNA_primer(n20up_primer_df,enzyme_df, enzyme_name,'',stype='n20up_primer_joint')\n",
    "\n",
    "\n",
    "     #质粒引物的设计类型：1---用户指定范围，2----无需用户指定范围，3----用户指定额外引物  \n",
    "    if plasmid_primer_desgin_type == 2: \n",
    "        n20down_primer_template = uha_dha_sgRNA_df[['Name','Region','n20_down_template','Target sequence','Rev Target sequence']]\n",
    "        n20down_primer_template['Region'] = n20down_primer_template['Name'] +';'+ n20down_primer_template['Region']\n",
    "        n20down_primer_df = p_d_seq.design_primer(n20down_primer_template,'Region','n20_down_template','sgRNA')\n",
    "        n20down_primer_df = pd.merge(n20down_primer_template[['Region','Target sequence','Rev Target sequence']],n20down_primer_df,on=['Region'],how='inner')\n",
    "        #加接头\n",
    "        n20down_primer_df = p_d_seq.add_joint_sgRNA_primer(n20down_primer_df,enzyme_df,enzyme_name,'',stype='n20down_primer_joint')\n",
    "\n",
    "    elif plasmid_primer_desgin_type == 1:\n",
    "        plasmid_backbone = promoter_terminator_down_terminator_seq\n",
    "        n20down_primer_df =  one_plasmid_system_design_by_user_region(uha_dha_sgRNA_df,seq_json,gb_path,plasmid_backbone,enzyme_df,enzyme_name)\n",
    "    elif plasmid_primer_desgin_type == 3:\n",
    "        plasmid_backbone = promoter_terminator_down_terminator_seq\n",
    "        primer_position_json, sgRNA_failture_primer = p_d_seq.check_locate_primer(plasmid_backbone, seq_json)\n",
    "        n20down_primer_df = one_plasmid_system_design_by_user_primer(seq_json,primer_position_json,gb_path,plasmid_backbone,enzyme_df,enzyme_name)\n",
    "\n",
    "\n",
    "    seq_altered_primer_template = info_df[info_df.seq_altered.apply(lambda x:len(x)>120)][['Name','Region','seq_altered']]\n",
    "    seq_altered_primer_template['Region'] = seq_altered_primer_template['Name'] +';'+ seq_altered_primer_template['Region']\n",
    "    seq_altered_primer_df = p_d_seq.design_primer(seq_altered_primer_template,'Region','seq_altered',stype='seq_altered')\n",
    "\n",
    "    #给同源臂引物加接头:uha取promoter_terminator_down_terminator_seq尾部反义4bp，dha取头promoter_terminator_up_promoter_seq正义4bp\n",
    "    uha_dha_primer_df = p_d_seq.add_joint_sgRNA_primer(uha_dha_info_primer_df,enzyme_df,enzyme_name,promoter_terminator_down_terminator_seq, promoter_terminator_up_promoter_seq, stype = 'u_d_primer_joint')\n",
    "\n",
    "    #seq_altered_primer加接头\n",
    "    seq_altered_primer_df = p_d_seq.add_joint_sgRNA_primer(seq_altered_primer_df,enzyme_df,enzyme_name,'',stype='seq_altered_primer_joint')\n",
    "\n",
    "    #分别提取，修饰后的uha、dha引物\n",
    "    in_col = ['Name','Region',\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",'product_value_joint','product_size_joint']\n",
    "    ou_col = ['Name','Region',\"u_primer_f_seq_(5'-3')\",\"u_primer_r_seq_(5'-3')\",'UHA','UHA_size',\"d_primer_f_seq_(5'-3')\",\"d_primer_r_seq_(5'-3')\",'DHA','DHA_size']\n",
    "    uha_primer_df, dha_primer_df = p_d_seq.create_uha_dha_primer_df(uha_dha_primer_df, in_col, ou_col)\n",
    "\n",
    "    n20down_primer_p_df = n20down_primer_df[[\"Region\",\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",'product_value_joint','product_size_joint']]\n",
    "    n20up_primer_p_df = n20up_primer_df[[\"Region\",\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",'product_value_joint','product_size_joint']]\n",
    "    seq_altered_primer_df = seq_altered_primer_df[[\"Region\",\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",'product_value_joint','product_size_joint']]\n",
    "\n",
    "    return uha_dha_sgRNA_df,uha_primer_df, dha_primer_df, n20down_primer_p_df, n20up_primer_p_df, seq_altered_primer_df,type_kind  \n",
    "\n",
    "#one plasmid sequencing primer  \n",
    "def one_plasmid_system_sequencing_design_primer(type_kind,uha_dha_sgRNA_df):\n",
    "    print(type_kind,\"1:代表启动子上游的序列大于600bp,同时终止子下游大于600bp\",\"2:代表启动子上游的序列小于600bp\",\"3:代表终止子上游序列小于600bp\")  \n",
    "\n",
    "    if type_kind == 3:\n",
    "        #载体测序\n",
    "        sequencing_primer_df=uha_dha_sgRNA_df[['Name', 'Region', 'UHA', 'UHA_size', 'DHA', 'DHA_size',\n",
    "                                    'plasmid', 'promoter_N20_terminator',\n",
    "                                    'promoter_N20_terminator_up', 'promoter_N20_terminator_down','seq_altered']]\n",
    "        sequencing_primer_df['plasmid_sequencing_region'] = sequencing_primer_df['UHA']+sequencing_primer_df['seq_altered']+sequencing_primer_df['DHA']+sequencing_primer_df['promoter_N20_terminator_up'] + sequencing_primer_df['promoter_N20_terminator']\n",
    "        sequencing_primer_df['Region'] = sequencing_primer_df['Name']+';'+sequencing_primer_df['Region']\n",
    "        sequencing_primer_template = sequencing_primer_df[['Name','Region','plasmid_sequencing_region','plasmid']]\n",
    "        plasmid_sequencing_primer_df = p_d_seq.create_sequencing_primer(sequencing_primer_template,sr,'plasmid','plasmid_sequencing_region',seq_type='plasmid_seq')\n",
    "\n",
    "    elif type_kind == 1:\n",
    "        #载体测序\n",
    "        #uha_dha测序\n",
    "        sequencing_primer_df=uha_dha_sgRNA_df[['Name', 'Region', 'UHA', 'UHA_size', 'DHA', 'DHA_size',\n",
    "                                    'plasmid', 'promoter_N20_terminator',\n",
    "                                    'promoter_N20_terminator_up', 'promoter_N20_terminator_down','seq_altered']]\n",
    "        sequencing_primer_df['plasmid_sequencing_region'] = sequencing_primer_df['UHA'] + sequencing_primer_df['seq_altered'] + sequencing_primer_df['DHA']\n",
    "        sequencing_primer_df['Region'] = sequencing_primer_df['Name']+';'+sequencing_primer_df['Region']\n",
    "        sequencing_primer_template = sequencing_primer_df[['Name','Region','plasmid_sequencing_region','plasmid']]\n",
    "        plasmid_sequencing_primer_df1 = p_d_seq.create_sequencing_primer(sequencing_primer_template,sr,'plasmid','plasmid_sequencing_region',seq_type='plasmid_seq')\n",
    "        #promoter_N20_terminator测序\n",
    "        sequencing_primer_df['plasmid_sequencing_region'] = sequencing_primer_df['promoter_N20_terminator']\n",
    "        sequencing_primer_template = sequencing_primer_df[['Name','Region','plasmid_sequencing_region','plasmid']]\n",
    "        plasmid_sequencing_primer_df2 = p_d_seq.create_sequencing_primer(sequencing_primer_template,sr,'plasmid','plasmid_sequencing_region',seq_type='plasmid_seq')\n",
    "        plasmid_sequencing_primer_df = p_d_seq.merge_sequencing_result(plasmid_sequencing_primer_df1, plasmid_sequencing_primer_df2)\n",
    "        plasmid_sequencing_primer_df2.to_csv('plasmid_sequencing_primer_df2.csv',index=False)\n",
    "    elif type_kind == 2:  \n",
    "        #载体测序  \n",
    "        #uha_dha测序\n",
    "        sequencing_primer_df=uha_dha_sgRNA_df[[ 'Name', 'Region', 'UHA', 'UHA_size', 'DHA', 'DHA_size',\n",
    "                                                'plasmid', 'promoter_N20_terminator',\n",
    "                                                'promoter_N20_terminator_up', 'promoter_N20_terminator_down','seq_altered']]\n",
    "        sequencing_primer_df['plasmid_sequencing_region'] = sequencing_primer_df['UHA'] + sequencing_primer_df['seq_altered'] + sequencing_primer_df['DHA'] + sequencing_primer_df['promoter_N20_terminator_up'] + sequencing_primer_df['promoter_N20_terminator'] \n",
    "        sequencing_primer_df['Region'] = sequencing_primer_df['Name']+';'+sequencing_primer_df['Region']\n",
    "        sequencing_primer_template = sequencing_primer_df[['Name','Region','plasmid_sequencing_region','plasmid']]\n",
    "        plasmid_sequencing_primer_df = p_d_seq.create_sequencing_primer(sequencing_primer_template,sr,'plasmid','plasmid_sequencing_region',seq_type='plasmid_seq')\n",
    "\n",
    "    temp = uha_dha_sgRNA_df[['Name','Region','UHA','DHA','promoter_N20_terminator','seq_altered']]\n",
    "    temp['Region'] = temp['Name']+';'+temp['Region']\n",
    "    temp = temp.drop(columns='Name')\n",
    "    sequencing_primer_template = pd.merge(sequencing_primer_template,temp,on=['Region'],how='inner')\n",
    "\n",
    "    return plasmid_sequencing_primer_df, sequencing_primer_template  \n",
    "\n",
    "def two_plasmid_system_n20_enzyme_cut_seq(no_ccdb_uha_dha_sgRNA_df,promoter_seq,enzyme_df,enzyme_name):\n",
    "\n",
    "\n",
    "    sgRNA_enzyme_df = enzyme_df[enzyme_df['name']==enzyme_name]\n",
    "    sgRNA_enzyme_df = sgRNA_enzyme_df.reset_index(drop=True)\n",
    "    cut_seq_len = sgRNA_enzyme_df.loc[0,'cut_seq_len']\n",
    "  \n",
    "\n",
    "    #N20片段引物退火获得\n",
    "    temp_sgRNA_df = no_ccdb_uha_dha_sgRNA_df[['Name', 'Region','Target sequence', 'promoter_N20_terminator']]\n",
    "    temp_sgRNA_df['Region'] = temp_sgRNA_df['Name'] +';'+ temp_sgRNA_df['Region']\n",
    "    temp_sgRNA_df.drop(columns='Name',inplace=True)\n",
    "    enzymeCutSeq_and_N20_df = p_d_seq.create_enzymeCutSeq_and_N20(temp_sgRNA_df, promoter_seq, enzyme_cut_len=cut_seq_len)\n",
    "    #取出重要列\n",
    "    enzymeCutSeq_and_N20_df = enzymeCutSeq_and_N20_df.rename(columns = {\"Region\":\"ID\"})\n",
    "    enzymeCutSeq_and_N20_df = enzymeCutSeq_and_N20_df[[ \"ID\",\"Target sequence\",\"enzymeCutSeq_and_N20\"  ]]\n",
    "\n",
    "    return enzymeCutSeq_and_N20_df\n",
    "\n",
    "def compute_distance_in_gb(gb_len,first,last):\n",
    "    if last < first:\n",
    "        seq_len = gb_len - first + last\n",
    "    else:\n",
    "        seq_len = last - first\n",
    "    return seq_len\n",
    "\n",
    "def region_2_distance(sgRNA_plasmid_seq_len, sgRNA_region_json, first_primer_start_position):\n",
    "\n",
    "    distance_dict = {}\n",
    "    for k,v in sgRNA_region_json.items():\n",
    "        arr = v.split(',')\n",
    "        first = int(arr[0])\n",
    "        last = int(arr[1]) \n",
    "        if first  > first_primer_start_position:\n",
    "            distance = first - first_primer_start_position\n",
    "        else:\n",
    "            distance = sgRNA_plasmid_seq_len - first_primer_start_position + first\n",
    "        distance_dict.update({distance:v})\n",
    "    sorted_distance = sorted(distance_dict.keys())\n",
    "\n",
    "    last_distance_dict = {}\n",
    "    for i in range(len(sorted_distance)):\n",
    "        v = distance_dict[sorted_distance[i]]\n",
    "        arr = v.split(',')\n",
    "        first = int(arr[0])\n",
    "        last = int(arr[1])\n",
    "        new_len = compute_distance_in_gb(sgRNA_plasmid_seq_len,first,last)\n",
    "        \n",
    "        if i != 0:\n",
    "            v = distance_dict[sorted_distance[i-1]]\n",
    "            arr = v.split(',')\n",
    "            first = int(arr[0])\n",
    "            last = int(arr[1])\n",
    "            old_len = compute_distance_in_gb(sgRNA_plasmid_seq_len,first,last)\n",
    "                \n",
    "            min_distance = sorted_distance[i] - sorted_distance[i-1] -old_len\n",
    "        else:\n",
    "            min_distance = sorted_distance[i]  \n",
    "\n",
    "        max_distance = min_distance + new_len\n",
    "\n",
    "        last_distance_dict.update({f'primer{i+1}':(min_distance, max_distance)})\n",
    "\n",
    "    print(last_distance_dict)\n",
    "\n",
    "    return last_distance_dict\n",
    "\n",
    "def two_plasmid_system_design_by_user_region(\n",
    "                                            no_ccdb_uha_dha_sgRNA_df,\n",
    "                                            ccdB_plasmid_backbone,\n",
    "                                            no_sgRNA_plasmid,\n",
    "                                            no_ccdb_plasmid,\n",
    "                                            uha_dha_sgRNA_df, \n",
    "                                            enzyme_df,\n",
    "                                            enzyme_name,\n",
    "                                            sgRNA_plasmid_seq,\n",
    "                                            sgRNA_plasmid_region_seq,\n",
    "                                            ccdb_plasmid_seq,\n",
    "                                            ccdb_plasmid_region_seq,\n",
    "                                            sgRNA_region_json,\n",
    "                                            ccdb_region_json\n",
    "                                            ):\n",
    "\n",
    "    # sgRNA_region_json\n",
    "    # ccdb_region_json\n",
    "        #区域转化距离    \n",
    "    # \"sgRNA_region_json\":{\n",
    "    #     \"region1\":\"371,570\",\n",
    "    #     \"region2\":\"3572,3770\"\n",
    "    # },\n",
    "    # distance_dict = {     \n",
    "    #         'distance1':(3000,3200),   \n",
    "    #         'distance2':(3000,3200),\n",
    "    #     }\n",
    "\n",
    "    if sgRNA_region_json != {}:\n",
    "        sgRNA_plasmid_primer_df = two_plasmid_system_design_by_user_sgRNA_region(uha_dha_sgRNA_df,sgRNA_plasmid_seq,sgRNA_plasmid_region_seq,sgRNA_region_json,no_ccdb_plasmid,enzyme_df,enzyme_name)\n",
    "    else:\n",
    "        sgRNA_plasmid_primer_df = two_plasmid_system_design_by_sgRNA_no_user(no_ccdb_uha_dha_sgRNA_df, uha_dha_sgRNA_df, enzyme_df, enzyme_name)\n",
    "\n",
    "    if ccdb_region_json != {}:\n",
    "        ccdb_plasmid_primer_df = two_plasmid_system_design_by_user_ccdb_region(uha_dha_sgRNA_df,no_sgRNA_plasmid,ccdb_plasmid_seq,ccdb_plasmid_region_seq,ccdb_region_json, enzyme_df, enzyme_name)\n",
    "    else:\n",
    "        ccdb_plasmid_primer_df = two_plasmid_system_design_by_ccdb_no_user(ccdB_plasmid_backbone,enzyme_df,enzyme_name)\n",
    "\n",
    "    return  sgRNA_plasmid_primer_df, ccdb_plasmid_primer_df   \n",
    "\n",
    "\n",
    "def two_plasmid_system_design_by_ccdb_no_user(ccdB_plasmid_backbone,enzyme_df,enzyme_name):\n",
    "        \n",
    "        #执行用户什么都不指定，设计ccdb质粒引物   \n",
    "        ccdb_primer_template_df = pd.DataFrame(columns=['plasmid_backbone_primer','plasmid_backbone'],data=[[f'ccdb_plasmid;primer',ccdB_plasmid_backbone]])\n",
    "        ccdb_plasmid_primer_df = p_d_seq.design_primer(ccdb_primer_template_df,'plasmid_backbone_primer','plasmid_backbone','plasmid')\n",
    "        #ccdb质粒引物加接头\n",
    "        ccdb_plasmid_primer_df = p_d_seq.add_joint_sgRNA_primer(ccdb_plasmid_primer_df,enzyme_df,enzyme_name,stype='ccdb_plasmid_primer_joint')\n",
    "        #提取引物的必要部分\n",
    "        ccdb_plasmid_primer_df = ccdb_plasmid_primer_df[['Region',r\"primer_f_seq_(5'-3')_joint\",r\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]    \n",
    "\n",
    "        return ccdb_plasmid_primer_df\n",
    "\n",
    "def two_plasmid_system_design_by_sgRNA_no_user(no_ccdb_uha_dha_sgRNA_df, uha_dha_sgRNA_df, enzyme_df, enzyme_name):\n",
    "        \n",
    "        #执行用户什么都不指定，设计sgRNA质粒引物\n",
    "        sgRNA_primer_template_df = no_ccdb_uha_dha_sgRNA_df[['Name','Region','sgRNA_template','Rev Target sequence']]\n",
    "        sgRNA_primer_template_df['Region'] = sgRNA_primer_template_df['Name'] +';'+ sgRNA_primer_template_df['Region']\n",
    "        sgRNA_plasmid_primer_df = p_d_seq.design_primer(sgRNA_primer_template_df,'Region','sgRNA_template','sgRNA')\n",
    "        uha_dha_sgRNA_df['Region'] = uha_dha_sgRNA_df['Name'] + ';' + uha_dha_sgRNA_df['Region']\n",
    "        #sgRNA质粒引物加接头\n",
    "        sgRNA_plasmid_primer_df = pd.merge(uha_dha_sgRNA_df[['Region','Target sequence','Rev Target sequence']],sgRNA_plasmid_primer_df,on=['Region'],how='inner')\n",
    "        sgRNA_plasmid_primer_df = p_d_seq.add_joint_sgRNA_primer(sgRNA_plasmid_primer_df,enzyme_df,enzyme_name,stype='sgRNA_plasmid_primer_joint')\n",
    "        #提取关键信息\n",
    "        sgRNA_plasmid_primer_df = sgRNA_plasmid_primer_df[['Region',r\"primer_f_seq_(5'-3')_joint\",r\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]\n",
    "\n",
    "        return sgRNA_plasmid_primer_df\n",
    "\n",
    "\n",
    "def two_plasmid_system_design_by_user_sgRNA_region(uha_dha_sgRNA_df,sgRNA_plasmid_seq,sgRNA_plasmid_region_seq,sgRNA_region_json,no_ccdb_plasmid,enzyme_df,enzyme_name):\n",
    "        \n",
    "        #sgRNA载体质粒                   #启动子终止子若横跨零点有问题\n",
    "        promoter_terminator_seq = sgRNA_plasmid_region_seq['promoter_seq'] + sgRNA_plasmid_region_seq['n20_coordinate_seq'] + sgRNA_plasmid_region_seq['terminator_seq']\n",
    "        promoter_terminator_start = sgRNA_plasmid_seq.find(promoter_terminator_seq)\n",
    "        promoter_terminator_end = promoter_terminator_start + len(promoter_terminator_seq)\n",
    "\n",
    "        first_primer_position_in_promoter_terminator = promoter_terminator_seq.find(sgRNA_plasmid_region_seq['terminator_seq'])\n",
    "        first_primer_start_position = promoter_terminator_start + first_primer_position_in_promoter_terminator\n",
    "        sgRNA_plasmid_seq_len = len(sgRNA_plasmid_seq)\n",
    "\n",
    "        sgRNA_distance_dict = region_2_distance(sgRNA_plasmid_seq_len, sgRNA_region_json,first_primer_start_position)\n",
    "        sgRNA_plasmid_primer_result_list = p_d_seq.design_primer_by_region_in_plasmid(first_primer_start_position, sgRNA_plasmid_seq, sgRNA_distance_dict, 20)\n",
    "        sgRNA_plasmid_primer_df =  su.result_primer_list_to_df(sgRNA_plasmid_primer_result_list)\n",
    "\n",
    "        #sgRNA质粒引物加接头\n",
    "        sgRNA_plasmid_primer_joint_df = p_d_seq.sgRNA_sgRNAprimer_merge(uha_dha_sgRNA_df, sgRNA_plasmid_primer_df)\n",
    "        sgRNA_primers_sum=len(sgRNA_plasmid_primer_df)\n",
    "        sgRNA_plasmid_primer_joint_df = p_d_seq.add_joint_plasmid_primer(enzyme_df,enzyme_name,sgRNA_plasmid_primer_joint_df,sgRNA_primers_sum,primer_type='sgRNA')\n",
    "        #修改df格式\n",
    "        sgRNA_plasmid_primer_df = p_d_seq.plasmid_primer(sgRNA_plasmid_primer_joint_df)\n",
    "        #添加索引\n",
    "        sgRNA_plasmid_primer_df.insert(0,'index',sgRNA_plasmid_primer_df.index)\n",
    "        sgRNA_plasmid_primer_df.reset_index(drop=True, inplace=True)\n",
    "        #给引物加产物\n",
    "        sgRNA_plasmid_primer_df = p_d_seq.add_product_and_size(no_ccdb_plasmid, sgRNA_plasmid_primer_df, enzyme_df, enzyme_name)\n",
    "\n",
    "        return sgRNA_plasmid_primer_df\n",
    "\n",
    "\n",
    "def two_plasmid_system_design_by_user_ccdb_region(uha_dha_sgRNA_df,no_sgRNA_plasmid,ccdb_plasmid_seq,ccdb_plasmid_region_seq,ccdb_region_json, enzyme_df, enzyme_name):\n",
    "\n",
    "        #ccdb载体质粒\n",
    "        ccdb_start = ccdb_plasmid_seq.find(ccdb_plasmid_region_seq['ccdb'])\n",
    "        ccdb_end = ccdb_start + len(ccdb_plasmid_region_seq['ccdb'])\n",
    "        first_primer_start_position = ccdb_end\n",
    "        ccdb_plasmid_seq_len = len(ccdb_plasmid_seq)\n",
    "        ccdb_distance_dict = region_2_distance(ccdb_plasmid_seq_len, ccdb_region_json,first_primer_start_position)\n",
    "        ccdb_plasmid_primer_result_list = p_d_seq.design_primer_by_region_in_plasmid(first_primer_start_position,ccdb_plasmid_seq, ccdb_distance_dict,len(ccdb_plasmid_region_seq['ccdb']))\n",
    "        ccdb_plasmid_primer_df = su.result_primer_list_to_df(ccdb_plasmid_primer_result_list)\n",
    "        #ccdB质粒引物加接头\n",
    "        ccdb_plasmid_primer_joint_df = p_d_seq.sgRNA_sgRNAprimer_merge(uha_dha_sgRNA_df,ccdb_plasmid_primer_df,sgRNA_columns=['Name','Region'])\n",
    "        ccdb_primers_sum = len(ccdb_plasmid_primer_df)\n",
    "        ccdb_plasmid_primer_joint_df = p_d_seq.add_joint_plasmid_primer(enzyme_df,enzyme_name,ccdb_plasmid_primer_joint_df,ccdb_primers_sum,primer_type='ccdb')\n",
    "        ccdb_plasmid_primer_joint_df =  ccdb_plasmid_primer_joint_df.drop(columns='Region').drop_duplicates()\n",
    "        #修改df格式\n",
    "        ccdb_plasmid_primer_df = p_d_seq.plasmid_primer(ccdb_plasmid_primer_joint_df)\n",
    "        #添加索引\n",
    "        ccdb_plasmid_primer_df.insert(0,'index',ccdb_plasmid_primer_df.index)\n",
    "        ccdb_plasmid_primer_df.reset_index(drop=True, inplace=True)\n",
    "        #给引物加产物\n",
    "        ccdb_plasmid_primer_df = p_d_seq.add_product_and_size(no_sgRNA_plasmid, ccdb_plasmid_primer_df, enzyme_df, enzyme_name=enzyme_name)\n",
    "\n",
    "        return ccdb_plasmid_primer_df\n",
    "\n",
    "\n",
    "def two_plasmid_system_design_by_no_user(no_ccdb_uha_dha_sgRNA_df,ccdB_plasmid_backbone,enzyme_df,enzyme_name):\n",
    "    #设计sgRNA质粒引物\n",
    "    no_ccdb_primer_template_df = no_ccdb_uha_dha_sgRNA_df[['Name','Region','sgRNA_template','Rev Target sequence']]\n",
    "    no_ccdb_primer_template_df['Region'] = no_ccdb_primer_template_df['Name'] + ';' + no_ccdb_primer_template_df['Region']\n",
    "    sgRNA_plasmid_primer_df = p_d_seq.design_primer(no_ccdb_primer_template_df,'Region','sgRNA_template','sgRNA')\n",
    "    no_ccdb_uha_dha_sgRNA_df['Region'] = no_ccdb_uha_dha_sgRNA_df['Name'] +';'+ no_ccdb_uha_dha_sgRNA_df['Region']\n",
    "    \n",
    "    #sgRNA质粒引物加接头\n",
    "    sgRNA_plasmid_primer_df = pd.merge(no_ccdb_uha_dha_sgRNA_df[['Region','Target sequence','Rev Target sequence']],sgRNA_plasmid_primer_df,on=['Region'],how='inner')\n",
    "    sgRNA_plasmid_primer_df = p_d_seq.add_joint_sgRNA_primer(sgRNA_plasmid_primer_df,enzyme_df,enzyme_name,stype='sgRNA_plasmid_primer_joint')\n",
    "    \n",
    "    #设计ccdB质粒引物\n",
    "    no_sgRNA_primer_template_df = pd.DataFrame(columns=['plasmid_backbone_primer','plasmid_backbone'],data=[[f'ccdb_plasmid;primer',ccdB_plasmid_backbone]])\n",
    "    ccdb_plasmid_primer_df = p_d_seq.design_primer(no_sgRNA_primer_template_df,'plasmid_backbone_primer','plasmid_backbone','plasmid')\n",
    "    \n",
    "    #ccdb质粒引物加接头\n",
    "    ccdb_plasmid_primer_df = p_d_seq.add_joint_sgRNA_primer(ccdb_plasmid_primer_df,enzyme_df,enzyme_name,stype='ccdb_plasmid_primer_joint')\n",
    "      \n",
    "    return sgRNA_plasmid_primer_df, ccdb_plasmid_primer_df\n",
    "\n",
    "def two_plasmid_system_design_by_user_primer_sgRNA( no_ccdb_plasmid,\n",
    "                                                    uha_dha_sgRNA_df,\n",
    "                                                    sgRNA_plasmid_backbone,\n",
    "                                                    sgRNA_promoter_terminator,\n",
    "                                                    sgRNA_primer_json,\n",
    "                                                    sgRNA_primer_position_json,\n",
    "                                                    enzyme_df,\n",
    "                                                    enzyme_name,\n",
    "                                                    n_20_label,\n",
    "                                                    ccdb_label,\n",
    "                                                    promoter_terminator_label):\n",
    "\n",
    "\n",
    "    #对引物进行排序, 确定所有引物\n",
    "    sgRNA_promoter_terminator_start = sgRNA_plasmid_backbone.find(sgRNA_promoter_terminator)\n",
    "    sgRNA_primer_dict_df, primers_sum = p_d_seq.sort_compose_primer(sgRNA_promoter_terminator_start,\n",
    "                                                         sgRNA_primer_json,\n",
    "                                                         sgRNA_primer_position_json,\n",
    "                                                         no_ccdb_plasmid,\n",
    "                                                         sgRNA_plasmid_backbone,\n",
    "                                                         n_20_label,\n",
    "                                                         ccdb_label,\n",
    "                                                         promoter_terminator_label)\n",
    "                                                         \n",
    "\n",
    "    #确定待合并的df\n",
    "    sgRNA_columns = ['Name','Region','Rev Target sequence']\n",
    "    temp_sgRNA_df = uha_dha_sgRNA_df[sgRNA_columns]\n",
    "    temp_sgRNA_df['Region'] = uha_dha_sgRNA_df['Name'] +';'+ uha_dha_sgRNA_df['Region']\n",
    "    temp_sgRNA_df.drop(columns = 'Name', inplace=True)\n",
    "    sgRNA_plasmid_primer_joint_df = su.merge_fill_two_df(temp_sgRNA_df, sgRNA_primer_dict_df)\n",
    "    #给引物加接头\n",
    "    sgRNA_plasmid_primer_joint_df = p_d_seq.add_joint_plasmid_primer(enzyme_df,enzyme_name,sgRNA_plasmid_primer_joint_df,int(primers_sum/2),primer_type='sgRNA')\n",
    "    #修改df格式\n",
    "    sgRNA_plasmid_primer_df = p_d_seq.plasmid_primer(sgRNA_plasmid_primer_joint_df)\n",
    "    #添加索引\n",
    "    sgRNA_plasmid_primer_df.insert(0,'index',sgRNA_plasmid_primer_df.index)\n",
    "    sgRNA_plasmid_primer_df.reset_index(drop=True, inplace=True)   \n",
    "    #给引物加产物\n",
    "    sgRNA_plasmid_primer_df = p_d_seq.add_product_and_size(no_ccdb_plasmid, sgRNA_plasmid_primer_df, enzyme_df, enzyme_name)\n",
    "    return sgRNA_plasmid_primer_df\n",
    "\n",
    "def two_plasmid_system_design_by_user_primer_ccdb(ccdB_plasmid_backbone, ccdb_primer_json, ccdb_primer_position_json, no_sgRNA_plasmid,enzyme_df,enzyme_name,n_20_label,ccdb_label,promoter_terminator_label):\n",
    "\n",
    "    \n",
    "    sgRNA_promoter_terminator_start = 0\n",
    "    #对引物进行排序,确定所有引物\n",
    "    ccdb_primer_dict_df, primers_sum = p_d_seq.sort_compose_primer(sgRNA_promoter_terminator_start,\n",
    "                                                         ccdb_primer_json,\n",
    "                                                         ccdb_primer_position_json,\n",
    "                                                         no_sgRNA_plasmid,\n",
    "                                                         ccdB_plasmid_backbone,\n",
    "                                                         n_20_label,\n",
    "                                                         ccdb_label,\n",
    "                                                         promoter_terminator_label)\n",
    "    #给引物加接头\n",
    "    ccdb_plasmid_primer_joint_df = p_d_seq.add_joint_plasmid_primer(enzyme_df,enzyme_name,\n",
    "                                                                 ccdb_primer_dict_df,\n",
    "                                                                 int(primers_sum/2),\n",
    "                                                                 primer_type='ccdb')\n",
    "    #修改df格式\n",
    "    ccdb_plasmid_primer_df = p_d_seq.plasmid_primer(ccdb_plasmid_primer_joint_df)\n",
    "    #添加索引\n",
    "    ccdb_plasmid_primer_df.insert(0,'index',ccdb_plasmid_primer_df.index)\n",
    "    ccdb_plasmid_primer_df.reset_index(drop=True, inplace=True)\n",
    "        \n",
    "    #给引物加产物\n",
    "    ccdb_plasmid_primer_df = p_d_seq.add_product_and_size(no_sgRNA_plasmid, ccdb_plasmid_primer_df, enzyme_df,enzyme_name)\n",
    "\n",
    "    return ccdb_plasmid_primer_df\n",
    "\n",
    "def two_plasmid_system_design_by_user_primer(\n",
    "                                            no_ccdb_uha_dha_sgRNA_df,\n",
    "                                            uha_dha_sgRNA_df,\n",
    "                                             sgRNA_primer_json, \n",
    "                                             ccdb_primer_json,\n",
    "                                            sgRNA_plasmid_backbone,\n",
    "                                            sgRNA_promoter_terminator,\n",
    "                                            ccdB_plasmid_backbone,\n",
    "                                            no_ccdb_plasmid,\n",
    "                                            no_sgRNA_plasmid,\n",
    "                                            enzyme_df,\n",
    "                                            enzyme_name,\n",
    "                                            n_20_label,\n",
    "                                            promoter_terminator_label,\n",
    "                                            ccdb_label):\n",
    "\n",
    "    #定位和检查引物\n",
    "    sgRNA_primer_position_json = {}\n",
    "    ccdb_primer_position_json = {}\n",
    "\n",
    "    if sgRNA_primer_json !={}:\n",
    "        sgRNA_primer_position_json, sgRNA_failture_primer = p_d_seq.check_locate_primer(sgRNA_plasmid_backbone, sgRNA_primer_json)\n",
    "    else:\n",
    "        sgRNA_plasmid_primer_df = two_plasmid_system_design_by_sgRNA_no_user(no_ccdb_uha_dha_sgRNA_df, uha_dha_sgRNA_df, enzyme_df, enzyme_name)\n",
    "        \n",
    "    if ccdb_primer_json !={}:  \n",
    "        ccdb_primer_position_json, ccdb_failture_primer = p_d_seq.check_locate_primer(ccdB_plasmid_backbone, ccdb_primer_json)\n",
    "    else:\n",
    "        ccdb_plasmid_primer_df = two_plasmid_system_design_by_ccdb_no_user(ccdB_plasmid_backbone,enzyme_df,enzyme_name)\n",
    "\n",
    "\n",
    "\n",
    "    if  sgRNA_primer_position_json == {} and sgRNA_primer_json !={}:\n",
    "        return sgRNA_failture_primer\n",
    "    elif sgRNA_primer_position_json !={} and sgRNA_primer_json != {}:\n",
    "        #设计sgRNA primer   \n",
    "        sgRNA_plasmid_primer_df = two_plasmid_system_design_by_user_primer_sgRNA( \n",
    "                                                                                no_ccdb_plasmid,\n",
    "                                                                                uha_dha_sgRNA_df,\n",
    "                                                                                sgRNA_plasmid_backbone,\n",
    "                                                                                sgRNA_promoter_terminator,\n",
    "                                                                                sgRNA_primer_json,\n",
    "                                                                                sgRNA_primer_position_json,\n",
    "                                                                                enzyme_df,\n",
    "                                                                                enzyme_name,\n",
    "                                                                                n_20_label,\n",
    "                                                                                ccdb_label,\n",
    "                                                                                promoter_terminator_label)\n",
    "    \n",
    "    if  ccdb_primer_position_json == {} and ccdb_primer_json !={}:\n",
    "        return ccdb_failture_primer\n",
    "    elif ccdb_primer_position_json != {} and ccdb_primer_json != {}:\n",
    "        #设计ccdb primer\n",
    "        ccdb_plasmid_primer_df = two_plasmid_system_design_by_user_primer_ccdb(ccdB_plasmid_backbone,\n",
    "                                                                                ccdb_primer_json,\n",
    "                                                                                ccdb_primer_position_json,\n",
    "                                                                                no_sgRNA_plasmid,\n",
    "                                                                                enzyme_df,\n",
    "                                                                                enzyme_name,\n",
    "                                                                                n_20_label,\n",
    "                                                                                ccdb_label,\n",
    "                                                                                promoter_terminator_label)\n",
    "              \n",
    "    return sgRNA_plasmid_primer_df, ccdb_plasmid_primer_df\n",
    "\n",
    "\n",
    "def two_plasmid_system_sequencing_design_primer(no_ccdb_uha_dha_sgRNA_df,no_sgRNA_uha_dha_ccdb_df):\n",
    "\n",
    "    #sgRNA质粒测序\n",
    "    sgRNA_plasmid_sequencing_primer_df = no_ccdb_uha_dha_sgRNA_df[['Name','Region','plasmid','promoter_N20_terminator','promoter_N20_terminator_up','promoter_N20_terminator_down','seq_altered']]\n",
    "    sgRNA_plasmid_sequencing_primer_df['plasmid_sequencing_region'] = sgRNA_plasmid_sequencing_primer_df['promoter_N20_terminator']\n",
    "    sgRNA_plasmid_sequencing_primer_df['Region'] = sgRNA_plasmid_sequencing_primer_df['Name']+';'+sgRNA_plasmid_sequencing_primer_df['Region']\n",
    "    sgRNA_plasmid_sequencing_primer_template = sgRNA_plasmid_sequencing_primer_df[['Name','Region','plasmid_sequencing_region','plasmid']]\n",
    "    sgRNA_plasmid_sequencing_primer_df = p_d_seq.create_sequencing_primer(sgRNA_plasmid_sequencing_primer_template,sr,'plasmid','plasmid_sequencing_region',seq_type='plasmid_seq')\n",
    "    #ccdb质粒载体测序\n",
    "    ccdb_plasmid_sequencing_primer_df = no_sgRNA_uha_dha_ccdb_df[['Name','Region','UHA','DHA','seq_altered','plasmid']]\n",
    "    ccdb_plasmid_sequencing_primer_df['plasmid_sequencing_region'] = ccdb_plasmid_sequencing_primer_df['UHA'] + ccdb_plasmid_sequencing_primer_df['seq_altered'] + ccdb_plasmid_sequencing_primer_df['DHA']\n",
    "    ccdb_plasmid_sequencing_primer_df['Region'] =  ccdb_plasmid_sequencing_primer_df['Name'] + ';' + ccdb_plasmid_sequencing_primer_df['Region']\n",
    "    ccdb_plasmid_sequencing_primer_template = ccdb_plasmid_sequencing_primer_df[['Name','Region','plasmid_sequencing_region','plasmid']]\n",
    "    ccdb_plasmid_sequencing_primer_df = p_d_seq.create_sequencing_primer(ccdb_plasmid_sequencing_primer_template,sr,'plasmid','plasmid_sequencing_region',seq_type='plasmid_seq')\n",
    "\n",
    "\n",
    "    sgRNA_temp = no_ccdb_uha_dha_sgRNA_df[['Name','Region','promoter_N20_terminator']]\n",
    "    sgRNA_temp['Region'] = sgRNA_temp['Name'] + ';' + sgRNA_temp['Region']\n",
    "    ccdb_temp = no_sgRNA_uha_dha_ccdb_df[['Name','Region','UHA','DHA','seq_altered']]\n",
    "    ccdb_temp['Region'] = ccdb_temp['Name'] + ';' + ccdb_temp['Region']\n",
    "    sgRNA_plasmid_sequencing_primer_template = pd.merge(sgRNA_plasmid_sequencing_primer_template, sgRNA_temp, on='Region', how='inner')\n",
    "    ccdb_plasmid_sequencing_primer_template = pd.merge(ccdb_plasmid_sequencing_primer_template, ccdb_temp,  on='Region', how='inner')\n",
    "\n",
    "    return sgRNA_plasmid_sequencing_primer_df,sgRNA_plasmid_sequencing_primer_template, ccdb_plasmid_sequencing_primer_df,ccdb_plasmid_sequencing_primer_template\n",
    "\n",
    "#genome sequencing primer\n",
    "def genome_sequencing_design_primer(info_input_df, uha_dha_df):\n",
    "\n",
    "    #编辑基因组测序\n",
    "    info_input_df1 = info_input_df[['name','region','seq_uha_max_whole','seq_dha_max_whole','uha_upstream','dha_downstream']].rename(columns={'name':'Name','region':'Region'})\n",
    "    UHA_DHA_df = pd.merge(info_input_df1,uha_dha_df)\n",
    "    UHA_DHA_df['sequencing_template'] = UHA_DHA_df['uha_upstream'] + UHA_DHA_df['seq_uha_max_whole']+UHA_DHA_df['seq_altered']+UHA_DHA_df['seq_dha_max_whole'] + UHA_DHA_df['dha_downstream'] \n",
    "    UHA_DHA_df['sequencing_region'] = UHA_DHA_df['UHA']+UHA_DHA_df['seq_altered']+UHA_DHA_df['DHA']\n",
    "    UHA_DHA_df['Region'] =  UHA_DHA_df['Name']+';'+UHA_DHA_df['Region']\n",
    "    genome_sequencing_primer_df = p_d_seq.create_sequencing_primer(UHA_DHA_df,sr,'sequencing_template','sequencing_region',seq_type='genome_seq')\n",
    "    #测序模板\n",
    "    genome_sequencing_template = UHA_DHA_df[['Region','sequencing_template','UHA','seq_altered','DHA']]\n",
    "\n",
    "    return genome_sequencing_primer_df, genome_sequencing_template\n",
    "  \n",
    "\n",
    "#     \n",
    "def execute_one_plasmid_system(plasmid_primer_desgin_type,\n",
    "                                sgRNA_region_seq_json,\n",
    "                                sgRNA_primer_json,\n",
    "                                gb_path,\n",
    "                                info_df,\n",
    "                                info_input_df,\n",
    "                                uha_dha_df,\n",
    "                                uha_dha_sgRNA_df,\n",
    "                                uha_dha_info_primer_df,\n",
    "                                uha_dha_primer_df,\n",
    "                                enzyme_df,  \n",
    "                                enzyme_name,\n",
    "                                output\n",
    "                                ):\n",
    "    gb = SeqIO.read(gb_path, \"genbank\")\n",
    "    gb_name = gb.name\n",
    "    \n",
    "    #质粒引物的设计类型：1---用户指定范围，2----无需用户指定范围，3----用户指定额外引物  \n",
    "    if plasmid_primer_desgin_type == 2:    \n",
    "        #无需用户指定范围\n",
    "        uha_dha_sgRNA_df,uha_primer_df,dha_primer_df,n20down_primer_p_df,n20up_primer_p_df, seq_altered_p_df,type_kind = one_plasmid_system_pcr_design_primer(                                                                                                                            \n",
    "                                                                                                                                        gb_path,\n",
    "                                                                                                                                        info_df,\n",
    "                                                                                                                                        uha_dha_sgRNA_df,\n",
    "                                                                                                                                        uha_dha_info_primer_df,\n",
    "                                                                                                                                        uha_dha_primer_df,\n",
    "                                                                                                                                        enzyme_df,\n",
    "                                                                                                                                        enzyme_name,\n",
    "                                                                                                                                        plasmid_primer_desgin_type,\n",
    "                                                                                                                                        sgRNA_region_seq_json\n",
    "                                                                                                                                        )\n",
    "        #修改 n20down_primer_p_df\n",
    "        temp = n20down_primer_p_df.loc[:0]\n",
    "        temp.loc[0,'Region'] = 1\n",
    "        n20down_primer_p_df = temp\n",
    "\n",
    "    elif plasmid_primer_desgin_type == 1:\n",
    "        #用户指定范围\n",
    "        uha_dha_sgRNA_df,uha_primer_df,dha_primer_df,n20down_primer_p_df,n20up_primer_p_df, seq_altered_p_df,type_kind = one_plasmid_system_pcr_design_primer(                                                                                                                            \n",
    "                                                                                                                                        gb_path,\n",
    "                                                                                                                                        info_df,\n",
    "                                                                                                                                        uha_dha_sgRNA_df,\n",
    "                                                                                                                                        uha_dha_info_primer_df,\n",
    "                                                                                                                                        uha_dha_primer_df,\n",
    "                                                                                                                                        enzyme_df,\n",
    "                                                                                                                                        enzyme_name,\n",
    "                                                                                                                                        plasmid_primer_desgin_type,\n",
    "                                                                                                                                        sgRNA_region_seq_json\n",
    "                                                                                                                                        )\n",
    "    elif plasmid_primer_desgin_type == 3:\n",
    "         #用户指定额外引物  \n",
    "        uha_dha_sgRNA_df,uha_primer_df,dha_primer_df,n20down_primer_p_df,n20up_primer_p_df, seq_altered_p_df,type_kind = one_plasmid_system_pcr_design_primer(                                                                                                                            \n",
    "                                                                                                                                        gb_path,\n",
    "                                                                                                                                        info_df,\n",
    "                                                                                                                                        uha_dha_sgRNA_df,\n",
    "                                                                                                                                        uha_dha_info_primer_df,\n",
    "                                                                                                                                        uha_dha_primer_df,\n",
    "                                                                                                                                        enzyme_df,\n",
    "                                                                                                                                        enzyme_name,\n",
    "                                                                                                                                        plasmid_primer_desgin_type,\n",
    "                                                                                                                                        sgRNA_primer_json\n",
    "                                                                                                                                        )\n",
    "\n",
    "    #设计质粒测序引物\n",
    "    plasmid_sequencing_primer_df, sequencing_primer_template = one_plasmid_system_sequencing_design_primer(type_kind,uha_dha_sgRNA_df)\n",
    "\n",
    " \n",
    "    #设计基因组测序引物\n",
    "    genome_sequencing_primer_df, genome_sequencing_template = genome_sequencing_design_primer(info_input_df, uha_dha_df)\n",
    "\n",
    "    #标准化，重命名   \n",
    "    df_common_list = su.rename_common_primer_df(n20up_primer_p_df, n20down_primer_p_df, seq_altered_p_df)\n",
    "    n20up_primer_p_df, n20down_primer_p_df, seq_altered_p_df = df_common_list[0], df_common_list[1], df_common_list[2]\n",
    "    uha_primer_df, dha_primer_df = su.rename_u_d_primer_df(uha_primer_df, dha_primer_df)\n",
    "    df_sequencing_list = su.rename_sequencing_primer_df(plasmid_sequencing_primer_df, genome_sequencing_primer_df)\n",
    "    plasmid_sequencing_primer_df, genome_sequencing_primer_df = df_sequencing_list[0], df_sequencing_list[1]\n",
    "\n",
    "    #----------------------------生成gb文件用于可视化展示-------------------------------------------------------------------\n",
    "    plasmid_primer_featrue_df = su.create_plasmid_primer_featrue_df(sequencing_primer_template,\n",
    "                                                                 uha_primer_df,\n",
    "                                                                 seq_altered_p_df,\n",
    "                                                                 dha_primer_df,   \n",
    "                                                                 n20up_primer_p_df)\n",
    "    plasmid_primer_featrue_df = plasmid_primer_featrue_df.fillna('')\n",
    "    joint_len, cut_seq_len = su.get_joint_by_enzyme(enzyme_df,enzyme_name)\n",
    "\n",
    "    #为每个编辑区域创建gb文件\n",
    "    pcr_gb_output = os.path.join(output,'one_plasmid_system_pcr_gb/')\n",
    "    if not exists(pcr_gb_output):\n",
    "        os.makedirs(pcr_gb_output)\n",
    "    pcr_tsv_df = p_d_seq.create_gb_for_region(plasmid_primer_featrue_df, n20down_primer_p_df, joint_len, cut_seq_len, pcr_gb_output, type='sgRNA_ccdb')\n",
    "\n",
    "\n",
    "    #质粒测序引物模板  \n",
    "    plasmid_sequencing_template = sequencing_primer_template\n",
    "    plasmid_sequencing_template.rename(columns={'Region':'ID','plasmid':'PLASMID','seq_altered':\"SEQ_ALTERED\",'promoter_N20_terminator':'PROMOTER_N20_TERMINATOR'},inplace=True)\n",
    "    plasmid_seq_gb_output = os.path.join(output,'one_plasmid_system_plasmid_sequencing_gb/')\n",
    "    if not exists(plasmid_seq_gb_output):\n",
    "        os.makedirs(plasmid_seq_gb_output)\n",
    "    plasmid_seq_tsv_df = p_d_seq.create_gb_for_sequencing_region(plasmid_sequencing_template, plasmid_sequencing_primer_df, plasmid_seq_gb_output, type='plasmid_sequencing')\n",
    "\n",
    "    #基因组测序引物模板\n",
    "    genome_sequencing_template.rename(columns={'Region':'ID','sequencing_template':'PLASMID','seq_altered':\"SEQ_ALTERED\"},inplace=True)\n",
    "    genome_seq_gb_output = os.path.join(output,'one_plasmid_system_genome_sequencing_gb/')\n",
    "    if not exists(genome_seq_gb_output):\n",
    "        os.makedirs(genome_seq_gb_output)\n",
    "\n",
    "    genome_seq_tsv_df = p_d_seq.create_gb_for_sequencing_region(genome_sequencing_template, genome_sequencing_primer_df, genome_seq_gb_output, type='genome_sequencing')\n",
    "\n",
    "    #合并三个df \n",
    "    tsv_df = pd.merge(pcr_tsv_df,plasmid_seq_tsv_df,on='name',how='inner').merge(genome_seq_tsv_df,on='name',how='inner')\n",
    "    tsv_df.to_csv(os.path.join(output,'one_plasmid_system_gb_visualization.tsv'), index=False, sep='\\t') \n",
    "    #-----------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "    #输出引物  \n",
    "    xlsx_file = os.path.join(\n",
    "        output,\n",
    "        'one_plasmid_design_result.xlsx'\n",
    "    )\n",
    "    with pd.ExcelWriter(xlsx_file) as writer:  \n",
    "        uha_primer_df.to_excel(writer,sheet_name = 'Primer_UHA',index_label='No.')\n",
    "        dha_primer_df.to_excel(writer,sheet_name = 'Primer_DHA',index_label='No.')\n",
    "        n20up_primer_p_df.to_excel(writer,sheet_name = 'Primer_inserted_fragment',index_label='No.')  \n",
    "        n20down_primer_p_df.to_excel(writer,sheet_name = 'Primer_plasmid',index_label='No.')\n",
    "        seq_altered_p_df.to_excel(writer,sheet_name = 'Seq_altered',index_label='No.')\n",
    "        plasmid_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_P1',index_label='No.')\n",
    "        genome_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_G',index_label='No.')\n",
    "\n",
    "    return xlsx_file                                                     \n",
    "\n",
    "def execute_two_plasmid_system(\n",
    "                                 method,\n",
    "                                 info_df,\n",
    "                                 uha_dha_info_primer_df,  \n",
    "                                 uha_dha_df,\n",
    "                                 info_input_df,\n",
    "                                 no_ccdb_plasmid,\n",
    "                                 no_sgRNA_plasmid,\n",
    "                                 uha_dha_sgRNA_df,\n",
    "                                 plasmid_primer_desgin_type,\n",
    "                                 enzyme_df,\n",
    "                                 enzyme_name,\n",
    "                                 sgRNA_primer_json,\n",
    "                                 ccdb_primer_json,  \n",
    "                                 sgRNA_region_seq_json,\n",
    "                                 ccdb_region_seq_json,\n",
    "                                 ccdb_label,\n",
    "                                 promoter_terminator_label,\n",
    "                                 n_20_label,\n",
    "                                 output\n",
    "                                 ):                                     \n",
    "\n",
    "    no_ccdb_uha_dha_sgRNA_df, sgRNA_plasmid_backbone, promoter_seq, terminator_seq, sgRNA_promoter_terminator = p_d_seq.create_new_plasmid(no_ccdb_plasmid, uha_dha_sgRNA_df.copy(), ccdb_label, promoter_terminator_label, n_20_label)\n",
    "    no_sgRNA_uha_dha_ccdb_df, ccdB_plasmid_backbone, ccdB_promoter_terminator_up_seq = p_d_seq.create_new_plasmid(no_sgRNA_plasmid, uha_dha_sgRNA_df.copy(), ccdb_label, promoter_terminator_label, n_20_label)\n",
    "    #酶切退火方式\n",
    "    \n",
    "\n",
    "    #质粒引物的设计类型：1---用户指定范围，2----无需用户指定范围，3----用户指定额外引物, \n",
    "    if  plasmid_primer_desgin_type == 1 and method == 'PCR':\n",
    "        #对质粒进行分割\n",
    "        #初步分割\n",
    "        sgRNA_plasmid_seq, sgRNA_plasmid_region_seq = p_d_seq.plasmid_region_division_by_labels(no_ccdb_plasmid, ccdb_label, promoter_terminator_label, n_20_label)\n",
    "        ccdb_plasmid_seq, ccdb_plasmid_region_seq = p_d_seq.plasmid_region_division_by_labels(no_sgRNA_plasmid, ccdb_label, promoter_terminator_label, n_20_label)\n",
    "\n",
    "        #第一段距离：sgRNA终止子距离第一段区域的最小、最大距离\n",
    "        #第二段距离：是第一段区域距离第二段区域的最小、最大距离\n",
    "        #。。。。。。。。。\n",
    "        #最后一段距离:最后一个区域距离sgRNA启动子距离\n",
    "\n",
    "\n",
    "        #用户确定区域，程序给你设计模板，保证引物的数量和质量\n",
    "        #将区域的序列转换成质粒上的坐标\n",
    "        \n",
    "        sgRNA_region_json = su.convert_seq_cor(no_ccdb_plasmid, sgRNA_region_seq_json)\n",
    "        ccdb_region_json = su.convert_seq_cor(no_sgRNA_plasmid, ccdb_region_seq_json)\n",
    "\n",
    "        #区域转换距离\n",
    "        # sgRNA_region_json={\n",
    "        #     \"region1\":\"371,570\",\n",
    "        #     \"region2\":\"3572,3770\"\n",
    "        # },\n",
    "        # ccdb_region_json={\n",
    "        #     \"region1\":\"8364,8563\",\n",
    "        #     \"region2\":\"376,575\"\n",
    "        # }\n",
    "        # distance_dict = {\n",
    "        #     'distance1':(3000,3200),\n",
    "        #     'distance2':(3000,3200),\n",
    "        # }\n",
    "        #根据区域设计引物\n",
    "        sgRNA_plasmid_primer_df, ccdb_plasmid_primer_df = two_plasmid_system_design_by_user_region(\n",
    "                                                                                    no_ccdb_uha_dha_sgRNA_df,\n",
    "                                                                                    ccdB_plasmid_backbone,\n",
    "                                                                                    no_sgRNA_plasmid,\n",
    "                                                                                    no_ccdb_plasmid,\n",
    "                                                                                    uha_dha_sgRNA_df,\n",
    "                                                                                    enzyme_df,\n",
    "                                                                                    enzyme_name,\n",
    "                                                                                    sgRNA_plasmid_seq,\n",
    "                                                                                    sgRNA_plasmid_region_seq,\n",
    "                                                                                    ccdb_plasmid_seq,\n",
    "                                                                                    ccdb_plasmid_region_seq,\n",
    "                                                                                    sgRNA_region_json,\n",
    "                                                                                    ccdb_region_json)\n",
    "\n",
    "    elif    plasmid_primer_desgin_type == 2 and method == 'PCR':\n",
    "            sgRNA_plasmid_primer_df, ccdb_plasmid_primer_df = two_plasmid_system_design_by_no_user(no_ccdb_uha_dha_sgRNA_df, ccdB_plasmid_backbone,enzyme_df,enzyme_name)\n",
    "    elif    plasmid_primer_desgin_type == 3 and method == 'PCR':\n",
    "         #用户提供正义链上的引物\n",
    "        # 引物 AACTATTTATCCAGTTGGTACAAAC\n",
    "        # sgRNA_primer_json = {\n",
    "        #     'primer3':'AACTATTTATCCAGTTGGTACAAAC',\n",
    "        #     'primer5':'GAAGATAACGAACAAAAACAATTGT'\n",
    "        # }\n",
    "        # ccdb_primer_json = {\n",
    "        #     'primer3':\"AACTGATTCAGTCTGATTTCGCGGT\",\n",
    "        #     'primer5':\"CCCTCTAATCGAAACTAATGGGGA\"\n",
    "        # }\n",
    "        #判断引物类型，设计引物\n",
    "        #sgRNA_primer\n",
    "        sgRNA_plasmid_primer_df, ccdb_plasmid_primer_df = two_plasmid_system_design_by_user_primer(\n",
    "                                                                                                    no_ccdb_uha_dha_sgRNA_df,\n",
    "                                                                                                    uha_dha_sgRNA_df,\n",
    "                                                                                                    sgRNA_primer_json, \n",
    "                                                                                                    ccdb_primer_json,\n",
    "                                                                                                    sgRNA_plasmid_backbone,\n",
    "                                                                                                    sgRNA_promoter_terminator,\n",
    "                                                                                                    ccdB_plasmid_backbone,\n",
    "                                                                                                    no_ccdb_plasmid,\n",
    "                                                                                                    no_sgRNA_plasmid,\n",
    "                                                                                                    enzyme_df,\n",
    "                                                                                                    enzyme_name,\n",
    "                                                                                                    n_20_label,\n",
    "                                                                                                    promoter_terminator_label,\n",
    "                                                                                                    ccdb_label)\n",
    "    elif    plasmid_primer_desgin_type == 1 and method == 'OLIGO':\n",
    "            #设计sgRNA质粒\n",
    "            enzymeCutSeq_and_N20_df = two_plasmid_system_n20_enzyme_cut_seq(no_ccdb_uha_dha_sgRNA_df, promoter_seq, enzyme_df, enzyme_name)\n",
    "\n",
    "            #设计ccdb质粒\n",
    "            ccdb_plasmid_seq, ccdb_plasmid_region_seq = p_d_seq.plasmid_region_division_by_labels(no_sgRNA_plasmid, ccdb_label, promoter_terminator_label, n_20_label)\n",
    "            ccdb_region_json = su.convert_seq_cor(no_sgRNA_plasmid, ccdb_region_seq_json)\n",
    "\n",
    "            ccdb_plasmid_primer_df = two_plasmid_system_design_by_user_ccdb_region(uha_dha_sgRNA_df,no_sgRNA_plasmid,ccdb_plasmid_seq,ccdb_plasmid_region_seq,ccdb_region_json, enzyme_df, enzyme_name)\n",
    "\n",
    "    elif    plasmid_primer_desgin_type == 2 and method == 'OLIGO':\n",
    "            #设计sgRNA质粒\n",
    "            enzymeCutSeq_and_N20_df = two_plasmid_system_n20_enzyme_cut_seq(no_ccdb_uha_dha_sgRNA_df, promoter_seq, enzyme_df, enzyme_name)\n",
    "            #设计ccdb质粒\n",
    "            ccdb_plasmid_primer_df = two_plasmid_system_design_by_ccdb_no_user(ccdB_plasmid_backbone,enzyme_df,enzyme_name)\n",
    "\n",
    "    elif    plasmid_primer_desgin_type == 3 and method == 'OLIGO':\n",
    "            #设计sgRNA质粒\n",
    "            enzymeCutSeq_and_N20_df = two_plasmid_system_n20_enzyme_cut_seq(no_ccdb_uha_dha_sgRNA_df, promoter_seq, enzyme_df, enzyme_name)\n",
    "            #设计ccdb质粒\n",
    "            ccdb_primer_position_json, ccdb_failture_primer = p_d_seq.check_locate_primer(ccdB_plasmid_backbone, ccdb_primer_json)\n",
    "\n",
    "            if ccdb_primer_position_json == {} and ccdb_primer_json != {}:\n",
    "                    return ccdb_failture_primer\n",
    "            \n",
    "            elif    ccdb_primer_position_json != {} and ccdb_primer_json != {}:\n",
    "                    #设计ccdb primer\n",
    "                    ccdb_plasmid_primer_df = two_plasmid_system_design_by_user_primer_ccdb(ccdB_plasmid_backbone,\n",
    "                                                                                ccdb_primer_json,\n",
    "                                                                                ccdb_primer_position_json,\n",
    "                                                                                no_sgRNA_plasmid,\n",
    "                                                                                enzyme_df,\n",
    "                                                                                enzyme_name,\n",
    "                                                                                n_20_label,\n",
    "                                                                                ccdb_label,\n",
    "                                                                                promoter_terminator_label)\n",
    "            \n",
    "\n",
    "    #设计seq_altered_primer，seq_altered > 120\n",
    "    seq_altered_primer_template =  info_df[info_df.seq_altered.apply(lambda x:len(x)>120)][['Name','Region','seq_altered']]\n",
    "    seq_altered_primer_template['Region'] = seq_altered_primer_template['Name'] +';'+ seq_altered_primer_template['Region']\n",
    "    seq_altered_primer_df = p_d_seq.design_primer(seq_altered_primer_template,'Region','seq_altered',stype='seq_altered')\n",
    "\n",
    "    #给引物加接头\n",
    "    #给同源臂引物加接头\n",
    "    uha_dha_primer_df = p_d_seq.add_joint_sgRNA_primer(uha_dha_info_primer_df, enzyme_df, enzyme_name, ccdB_plasmid_backbone, stype = 'u_d_primer_joint')\n",
    "    #seq_altered_primer加接头\n",
    "    seq_altered_primer_df = p_d_seq.add_joint_sgRNA_primer(seq_altered_primer_df, enzyme_df, enzyme_name, stype='seq_altered_primer_joint')\n",
    "\n",
    "    #分别提取，修饰后的uha、dha引物\n",
    "    in_col = ['Name','Region',\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",'product_value_joint','product_size_joint']\n",
    "    ou_col = ['Name','Region',\"u_primer_f_seq_(5'-3')\",\"u_primer_r_seq_(5'-3')\",'UHA','UHA_size',\"d_primer_f_seq_(5'-3')\",\"d_primer_r_seq_(5'-3')\",'DHA','DHA_size']\n",
    "    uha_primer_df, dha_primer_df = p_d_seq.create_uha_dha_primer_df(uha_dha_primer_df, in_col, ou_col)\n",
    "    #提取变化序列引物\n",
    "    seq_altered_p_df = seq_altered_primer_df[['Region',\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]\n",
    "\n",
    "    \n",
    "\n",
    " \n",
    "\n",
    "    # if 'index' in sgRNA_plasmid_primer_df.columns and 'index' in ccdb_plasmid_primer_df.columns:\n",
    "       \n",
    "    #     sgRNA_plasmid_p_df = sgRNA_plasmid_primer_df\n",
    "    #     ccdb_plasmid_p_df = ccdb_plasmid_primer_df\n",
    "\n",
    "    #提取引物的必要部分\n",
    "    if method == 'PCR':\n",
    "        sgRNA_plasmid_p_df = sgRNA_plasmid_primer_df[['Region',r\"primer_f_seq_(5'-3')_joint\",r\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]\n",
    "    ccdb_plasmid_p_df = ccdb_plasmid_primer_df[['Region',r\"primer_f_seq_(5'-3')_joint\",r\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]\n",
    "\n",
    "    \n",
    "    # else:\n",
    "    #     if method =='PCR':\n",
    "    #         sgRNA_plasmid_p_df = sgRNA_plasmid_primer_df[['Region',\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]\n",
    "    #         ccdb_plasmid_p_df = ccdb_plasmid_primer_df[['Region',\"primer_f_seq_(5'-3')_joint\",\"primer_r_seq_(5'-3')_joint\",\"product_value_joint\",\"product_size_joint\"]]\n",
    "\n",
    "    #设计质粒测序引物\n",
    "    sgRNA_plasmid_sequencing_primer_df,sgRNA_plasmid_sequencing_primer_template, ccdb_plasmid_sequencing_primer_df,ccdb_plasmid_sequencing_primer_template = two_plasmid_system_sequencing_design_primer(no_ccdb_uha_dha_sgRNA_df,no_sgRNA_uha_dha_ccdb_df)\n",
    "\n",
    "    #设计基因组测序引物\n",
    "    genome_sequencing_primer_df,genome_sequencing_template = genome_sequencing_design_primer(info_input_df, uha_dha_df)\n",
    "\n",
    "    #标准化，重命名\n",
    "    if method == 'PCR':\n",
    "        df_common_list = su.rename_common_primer_df(sgRNA_plasmid_p_df,ccdb_plasmid_p_df,seq_altered_p_df)\n",
    "        sgRNA_plasmid_p_df, ccdb_plasmid_p_df, seq_altered_p_df = df_common_list[0], df_common_list[1], df_common_list[2]\n",
    "    else:\n",
    "        df_common_list = su.rename_common_primer_df(ccdb_plasmid_p_df,seq_altered_p_df)\n",
    "        ccdb_plasmid_p_df, seq_altered_p_df = df_common_list[0], df_common_list[1]\n",
    "\n",
    "    uha_primer_df, dha_primer_df = su.rename_u_d_primer_df(uha_primer_df,dha_primer_df)\n",
    "    df_sequencing_list = su.rename_sequencing_primer_df(sgRNA_plasmid_sequencing_primer_df, ccdb_plasmid_sequencing_primer_df,genome_sequencing_primer_df)\n",
    "    sgRNA_plasmid_sequencing_primer_df, ccdb_plasmid_sequencing_primer_df, genome_sequencing_primer_df = df_sequencing_list[0], df_sequencing_list[1], df_sequencing_list[2]\n",
    "\n",
    "    \n",
    "    \n",
    "    #------------------------------------------生成gb文件用于引物的可视化展示-------------------------------------------------------------\n",
    "    #--------------------------PCR----------------------生成sgRNA_gb文件-------------------------------------------------------------------\n",
    "    sgRNA_plasmid_sequencing_primer_template = sgRNA_plasmid_sequencing_primer_template[['Region','plasmid','promoter_N20_terminator']].rename(columns={'Region':'ID','plasmid':\"PLASMID\",\"promoter_N20_terminator\":\"PROMOTER_N20_TERMINATOR\"})\n",
    "    if method == 'PCR':\n",
    "        sgRNA_plasmid_primer = sgRNA_plasmid_p_df[['ID', 'PRIMER_LEFT_WHOLE_SEQUENCE', 'PRIMER_RIGHT_WHOLE_SEQUENCE']]\n",
    "    else:\n",
    "        sgRNA_plasmid_primer = enzymeCutSeq_and_N20_df[['ID','Target sequence']]\n",
    "    joint_len, cut_seq_len = su.get_joint_by_enzyme(enzyme_df,enzyme_name)\n",
    "    \n",
    "    #为每个编辑区域创建gb文件\n",
    "    gb_output = os.path.join(output,'two_plasmid_system_pcr_gb/')\n",
    "    if not exists(gb_output):\n",
    "        os.makedirs(gb_output)   \n",
    "    if method == 'PCR':\n",
    "        type = 'sgRNA'\n",
    "        sgRNA_pcr_tsv_df = p_d_seq.create_gb_for_region(sgRNA_plasmid_sequencing_primer_template, sgRNA_plasmid_primer, joint_len, cut_seq_len, gb_output,type)\n",
    "    elif method == 'OLIGO':\n",
    "        type = 'enzyme_cut'\n",
    "        sgRNA_pcr_tsv_df = p_d_seq.create_gb_for_region(sgRNA_plasmid_sequencing_primer_template, sgRNA_plasmid_primer, joint_len, cut_seq_len, gb_output,type)\n",
    "\n",
    "\n",
    "    #---------------------------PCR---------------------生成ccdb_gb文件---------------------------------------------------------------------\n",
    "    # plasmid_primer_featrue_df = ccdb_plasmid_sequencing_primer_template[['Region','plasmid']].rename(columns={'Region':'ID','plasmid':\"PLASMID\"})\n",
    "    # [['Name','Region','UHA','DHA','seq_altered']]\n",
    "    plasmid_primer_featrue_df = su.create_plasmid_primer_featrue_df(ccdb_plasmid_sequencing_primer_template,\n",
    "                                                                 uha_primer_df,\n",
    "                                                                 seq_altered_p_df,\n",
    "                                                                 dha_primer_df)\n",
    "    plasmid_primer_featrue_df = plasmid_primer_featrue_df.fillna('')\n",
    "    joint_len, cut_seq_len = su.get_joint_by_enzyme(enzyme_df,enzyme_name)\n",
    "\n",
    "    #为每个编辑区域创建gb文件\n",
    "    gb_output = os.path.join(output ,'two_plasmid_system_pcr_gb/')\n",
    "    if not exists(gb_output):\n",
    "        os.makedirs(gb_output)   \n",
    "    ccdb_pcr_tsv_df = p_d_seq.create_gb_for_region(plasmid_primer_featrue_df, ccdb_plasmid_p_df,joint_len, cut_seq_len, gb_output,type='ccdb')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #---------------------plasmid-------SEQUENCING------------------------------------------------------------------------------------------------------------\n",
    "    #为每个编辑区域创建gb文件\n",
    "    gb_output = os.path.join(output,'two_plasmid_system_plasmid_sequencing_gb/')\n",
    "    if not exists(gb_output):\n",
    "        os.makedirs(gb_output)\n",
    "    sgRNA_plasmid_primer_featrue_df = sgRNA_plasmid_sequencing_primer_template\n",
    "    sgRNA_plasmid_sequencing_tsv_df = p_d_seq.create_gb_for_sequencing_region(sgRNA_plasmid_primer_featrue_df, sgRNA_plasmid_sequencing_primer_df, gb_output, type='sgRNA_plasmid_sequencing')\n",
    "\n",
    "\n",
    "    ccdb_plasmid_primer_featrue_df = ccdb_plasmid_sequencing_primer_template[['Region','plasmid','UHA','DHA']].rename(columns={'Region':'ID','plasmid':\"PLASMID\"})\n",
    "    ccdb_plasmid_sequencing_tsv_df = p_d_seq.create_gb_for_sequencing_region(ccdb_plasmid_primer_featrue_df, ccdb_plasmid_sequencing_primer_df, gb_output, type='ccdb_plasmid_sequencing')\n",
    "\n",
    "    #-------------------genome----------SEQUENCING------------------------------------------------------------------\n",
    "    gb_output = os.path.join(output,'two_plasmid_system_genome_sequencing_gb/')\n",
    "    if not exists(gb_output):\n",
    "        os.makedirs(gb_output)\n",
    "    genome_sequencing_template.rename(columns={'Region':'ID','sequencing_template':'PLASMID','seq_altered':\"SEQ_ALTERED\"},inplace=True)\n",
    "    genome_sequencing_tsv_df = p_d_seq.create_gb_for_sequencing_region(genome_sequencing_template, genome_sequencing_primer_df, gb_output, type='genome_sequencing')\n",
    "\n",
    "\n",
    "    #生成tsv\n",
    "    # print(sgRNA_tsv_df,'\\n',ccdb_tsv_df)  \n",
    "    pcr_df = pd.merge(sgRNA_pcr_tsv_df, ccdb_pcr_tsv_df, on='name')  \n",
    "    sequencing_df = pd.merge(sgRNA_plasmid_sequencing_tsv_df, ccdb_plasmid_sequencing_tsv_df,on='name')\n",
    "    tsv_df = pcr_df.merge(sequencing_df,on='name').merge(genome_sequencing_tsv_df,on='name')\n",
    "    tsv_df.to_csv(os.path.join(output,'two_plasmid_system_gb_visualization.tsv'), index=False, sep='\\t')\n",
    "    #-------------------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "    #输出引物     \n",
    "    xlsx_file = os.path.join(\n",
    "        output,\n",
    "        'two_plasmid_design_result.xlsx'\n",
    "    )\n",
    "    if method == 'PCR':\n",
    "        with pd.ExcelWriter(xlsx_file) as writer:  \n",
    "            uha_primer_df.to_excel(writer,sheet_name = 'Primer_UHA',index_label='No.')\n",
    "            dha_primer_df.to_excel(writer,sheet_name = 'Primer_DHA',index_label='No.')\n",
    "            sgRNA_plasmid_p_df.to_excel(writer,sheet_name = 'sgRNA_plasmid_fragment',index_label='No.')  \n",
    "            ccdb_plasmid_p_df.to_excel(writer,sheet_name = 'uha_dha_plasmid',index_label='No.')\n",
    "            seq_altered_p_df.to_excel(writer,sheet_name = 'Seq_altered',index_label='No.')\n",
    "            sgRNA_plasmid_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_P1',index_label='No.')\n",
    "            ccdb_plasmid_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_P2',index_label='No.')\n",
    "            genome_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_G',index_label='No.')\n",
    "    else:\n",
    "         with pd.ExcelWriter(xlsx_file) as writer:  \n",
    "            uha_primer_df.to_excel(writer,sheet_name = 'Primer_UHA',index_label='No.')\n",
    "            dha_primer_df.to_excel(writer,sheet_name = 'Primer_DHA',index_label='No.')\n",
    "            enzymeCutSeq_and_N20_df.to_excel(writer,sheet_name = 'sgRNA_plasmid_fragment',index_label='No.')  \n",
    "            ccdb_plasmid_p_df.to_excel(writer,sheet_name = 'uha_dha_plasmid',index_label='No.')\n",
    "            seq_altered_p_df.to_excel(writer,sheet_name = 'Seq_altered',index_label='No.')\n",
    "            sgRNA_plasmid_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_P1',index_label='No.')\n",
    "            ccdb_plasmid_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_P2',index_label='No.')\n",
    "            genome_sequencing_primer_df.to_excel(writer,sheet_name = 'Test_primer_G',index_label='No.')\n",
    "\n",
    "    return xlsx_file\n",
    "\n",
    "  \n",
    "def read_chopchopInput_add_uha_dha(genome_path,chopchop_input,uha_dha_params):\n",
    "    max_left_arm_seq_length = uha_dha_params['max_left_arm_seq_length']\n",
    "    max_right_arm_seq_length = uha_dha_params['max_right_arm_seq_length']\n",
    "\n",
    "    info_input_df = su.del_Unnamed(pd.read_csv(chopchop_input))\n",
    "\n",
    "    info_input_df.columns = [i.lower() for i in info_input_df.columns]\n",
    "\n",
    "\n",
    "    def work(mun_id,geneid, mutation_pos_index):\n",
    "        if mutation_pos_index - max_left_arm_seq_length < 0:\n",
    "            error_message = \"The length of upstream sequence of manipulation site of \" + mun_id + \" must be larger than sum of 'Max Length of UHA' and 'Max Length of UIS'.\"\n",
    "            return error_message,error_message,error_message,error_message\n",
    "\n",
    "        record = su.extract_seq_from_genome(genome_path,geneid)\n",
    "\n",
    "        seq_uha_max_whole = str(record[\n",
    "                        mutation_pos_index - max_left_arm_seq_length : mutation_pos_index\n",
    "                        ])\n",
    "        seq_dha_max_whole = str(record[\n",
    "                                mutation_pos_index : mutation_pos_index + max_right_arm_seq_length\n",
    "                                ])\n",
    "\n",
    "\n",
    "        uha_upstream = str(  \n",
    "                        record[\n",
    "                            mutation_pos_index - max_left_arm_seq_length - 100 : mutation_pos_index - max_left_arm_seq_length\n",
    "                        ]\n",
    "                    )\n",
    "        dha_downstream=str(\n",
    "                        record[\n",
    "                            mutation_pos_index + max_right_arm_seq_length  : mutation_pos_index + max_right_arm_seq_length  + 100\n",
    "                        ]\n",
    "                    )\n",
    "        return  uha_upstream, dha_downstream, seq_uha_max_whole, seq_dha_max_whole\n",
    "\n",
    "    info_df = su.lambda2cols(info_input_df,lambdaf=work,in_coln=['name','geneid','mutation_pos_index'],to_colns=['uha_upstream','dha_downstream','seq_uha_max_whole','seq_dha_max_whole'])\n",
    "\n",
    "    return info_df  \n",
    "\n",
    "def check_quality_control(plasmid_backbone,seq_json):\n",
    "    \n",
    "    failture_seq_json = {}\n",
    "    print(failture_seq_json == {})  \n",
    "    for k,v in seq_json.items():\n",
    "        plasmid_backbone = plasmid_backbone.upper()\n",
    "        start = plasmid_backbone.find(v.upper())\n",
    "        errorMessage = \"\"\n",
    "\n",
    "\n",
    "         #是否是DNA序列\n",
    "        if not su.is_dna(v):\n",
    "            errorMessage = f\"{v}: The sequence is not a DNA sequence\" \n",
    "            failture_seq_json.update({k:errorMessage})\n",
    "            raise ValueError(errorMessage)  \n",
    "\n",
    "        if start == -1:\n",
    "            failture_seq_json.update({k: f'{v}:The sequence is not on the plasmid, please enter the sequence on the sense chain of the plasmid'})\n",
    "            errorMessage = f'{v}: The sequence is not on the plasmid, please enter the sequence on the sense chain of the plasmid'\n",
    "            failture_seq_json.update({k:errorMessage})\n",
    "            raise ValueError(errorMessage)  \n",
    "        else:\n",
    "            #判断 5 < 序列长度 < 300\n",
    "            if len(v) < 5  or  len(v) > 300:\n",
    "                errorMessage = f\"{v}: The sequence length should be between 5 and 300\" \n",
    "                failture_seq_json.update({k:errorMessage}) \n",
    "                raise ValueError(errorMessage)  \n",
    "           \n",
    "            # 去取质粒骨架，引物的位置距离质粒骨架起始的距离大于50bp，引物的位置距离质粒骨架终止的距离大于50bp\n",
    "            if start < 50 or len(plasmid_backbone) - (start + len(v)) < 50:\n",
    "                errorMessage = f'{v} The distance between the position of the sequence and the start of the plasmid skeleton should be greater than 50 bp, and the distance between the position of the sequence and the end of the plasmid skeleton should be greater than 50 bp'\n",
    "                failture_seq_json.update({k:errorMessage})\n",
    "                raise ValueError(errorMessage)  \n",
    "        \n",
    "\n",
    "    return failture_seq_json\n",
    "   \n",
    "def check_plasmid(gb_path, ccdb_label='', promoter_terminator_label='', n_20_label=''):\n",
    "\n",
    "    gb = SeqIO.read(gb_path, \"genbank\")\n",
    "    gb_seq = str(gb.seq)\n",
    "    #get coordinate\n",
    "    ccdb_coordinate = su.get_feature_coordinate(ccdb_label,gb_path) \n",
    "    #N20\n",
    "    n20_coordinate = su.get_feature_coordinate(n_20_label,gb_path)\n",
    "\n",
    "    if ccdb_coordinate != (-1,-1) and n20_coordinate !=(-1,-1):\n",
    "        return 'one_plasmid_file_path'\n",
    "    elif ccdb_coordinate==(-1,-1) and n20_coordinate !=(-1,-1):\n",
    "        return 'no_ccdb_plasmid'\n",
    "    elif ccdb_coordinate != (-1,-1) and n20_coordinate ==(-1,-1):\n",
    "        return 'no_sgRNA_plasmid'\n",
    "    else:\n",
    "        raise ValueError('The plasmid you uploaded does not contain the necessary tags')\n",
    "\n",
    "\n",
    "def check_enzyme(enzyme,enzyme_df):\n",
    "    name = enzyme['enzyme_name']\n",
    "    gap_seq = enzyme['gap_sequence']\n",
    "    protection_seq = enzyme['protection_sequence']\n",
    "    errorMessage = \"\" \n",
    "\n",
    "    if len(gap_seq) < 1 or len(gap_seq) >10:\n",
    "        errorMessage = f\"The gap sequence:{gap_seq} length should be greater than 1 and less than 10\" \n",
    "        raise ValueError(errorMessage) \n",
    "    if not su.is_dna(gap_seq):\n",
    "        errorMessage = f'The gap sequence:{gap_seq} is not a DNA sequence'\n",
    "        raise ValueError(errorMessage) \n",
    "    if not su.is_dna(protection_seq):\n",
    "        errorMessage = f'The protection sequence:{gap_seq} is not a DNA sequence'\n",
    "        raise ValueError(errorMessage) \n",
    "    if len(protection_seq) < 1 or len(protection_seq) >20:\n",
    "        errorMessage = f\"The protection sequence:{gap_seq} length should be greater than 1 and less than 10\" \n",
    "        raise ValueError(errorMessage) \n",
    "    \n",
    "    if name in list(enzyme_df['name']) and len(enzyme_df[enzyme_df['name']==name]) == 1:\n",
    "        enzyme_df = enzyme_df[enzyme_df['name']==name]\n",
    "        enzyme_df['protective_base'] = protection_seq\n",
    "        enzyme_df['gap_len'] = len(gap_seq)\n",
    "        enzyme_df['gap_seq'] = gap_seq   \n",
    "        return enzyme_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7639f129",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "48369f97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/yanghe/program/edit_sequence_design\n",
      "-1 -1\n",
      "5057 5363\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "9002 9022\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "8696 8716\n",
      "-1 -1\n",
      "5057 5363\n",
      "-1 -1\n",
      "-1 -1\n",
      "-1 -1\n",
      "执行方法： PCR\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 132, not in any ok right region 172, ok 104', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 10.796287109339062, 'PRIMER_LEFT_0_PENALTY': 3.4730231907085454, 'PRIMER_RIGHT_0_PENALTY': 7.323263918630516, 'PRIMER_LEFT_0_SEQUENCE': 'GTTATTTCCGTGCGCGTTGGTGA', 'PRIMER_RIGHT_0_SEQUENCE': 'GCCTGCATTTATAAAGTGCTGACCC', 'PRIMER_LEFT_0': (0, 23), 'PRIMER_RIGHT_0': (1038, 25), 'PRIMER_LEFT_0_TM': 64.52697680929145, 'PRIMER_RIGHT_0_TM': 62.676736081369484, 'PRIMER_LEFT_0_GC_PERCENT': 52.17391304347826, 'PRIMER_RIGHT_0_GC_PERCENT': 48.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 19.610968059260017, 'PRIMER_RIGHT_0_SELF_ANY_TH': 24.27068115227172, 'PRIMER_LEFT_0_SELF_END_TH': 5.11952728907761, 'PRIMER_RIGHT_0_SELF_END_TH': 15.625926837095221, 'PRIMER_LEFT_0_HAIRPIN_TH': 36.149651650343515, 'PRIMER_RIGHT_0_HAIRPIN_TH': 39.3011119927678, 'PRIMER_LEFT_0_END_STABILITY': 4.02, 'PRIMER_RIGHT_0_END_STABILITY': 4.46, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1039}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 22, not in any ok right region 172, ok 214', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 3.078270799216739, 'PRIMER_LEFT_0_PENALTY': 3.0595622503428217, 'PRIMER_RIGHT_0_PENALTY': 0.018708548873917152, 'PRIMER_LEFT_0_SEQUENCE': 'CAAAGTGGTGGTGATGCGCTTCG', 'PRIMER_RIGHT_0_SEQUENCE': 'TGATCCTGCCGGTGAAGCGG', 'PRIMER_LEFT_0': (0, 23), 'PRIMER_RIGHT_0': (1028, 20), 'PRIMER_LEFT_0_TM': 65.05956225034282, 'PRIMER_RIGHT_0_TM': 64.98129145112608, 'PRIMER_LEFT_0_GC_PERCENT': 56.52173913043478, 'PRIMER_RIGHT_0_GC_PERCENT': 65.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 17.329262194548335, 'PRIMER_RIGHT_0_SELF_ANY_TH': 16.30857360756403, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 16.30857360756403, 'PRIMER_LEFT_0_HAIRPIN_TH': 0.0, 'PRIMER_RIGHT_0_HAIRPIN_TH': 36.32007900898992, 'PRIMER_LEFT_0_END_STABILITY': 3.79, 'PRIMER_RIGHT_0_END_STABILITY': 5.52, 'PRIMER_PAIR_0_COMPL_ANY_TH': 16.088027256334385, 'PRIMER_PAIR_0_COMPL_END_TH': 3.262359280960652, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1029}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, low tm 2, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, high hairpin stability 3, not in any ok right region 172, ok 233', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 4.671901001554829, 'PRIMER_LEFT_0_PENALTY': 4.51246984845892, 'PRIMER_RIGHT_0_PENALTY': 0.15943115309590894, 'PRIMER_LEFT_0_SEQUENCE': 'CCTGAAATTTCCGTCATGCCGACC', 'PRIMER_RIGHT_0_SEQUENCE': 'CTGCGTGCAGGTCCCTGGTT', 'PRIMER_LEFT_0': (0, 24), 'PRIMER_RIGHT_0': (1022, 20), 'PRIMER_LEFT_0_TM': 64.48753015154108, 'PRIMER_RIGHT_0_TM': 65.15943115309591, 'PRIMER_LEFT_0_GC_PERCENT': 54.166666666666664, 'PRIMER_RIGHT_0_GC_PERCENT': 65.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 15.483905306687745, 'PRIMER_RIGHT_0_SELF_ANY_TH': 13.998505944661474, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 0.0, 'PRIMER_LEFT_0_HAIRPIN_TH': 42.07251419716505, 'PRIMER_RIGHT_0_HAIRPIN_TH': 43.77426119223003, 'PRIMER_LEFT_0_END_STABILITY': 4.79, 'PRIMER_RIGHT_0_END_STABILITY': 3.67, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1023}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 1, high hairpin stability 57, not in any ok right region 172, ok 178', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 7.3210989619772135, 'PRIMER_LEFT_0_PENALTY': 4.05089504894022, 'PRIMER_RIGHT_0_PENALTY': 3.270203913036994, 'PRIMER_LEFT_0_SEQUENCE': 'GACTTCTTCTTCGCGCCCAA', 'PRIMER_RIGHT_0_SEQUENCE': 'CGAAGCAAGCGCGTGTAGATGA', 'PRIMER_LEFT_0': (0, 20), 'PRIMER_RIGHT_0': (1024, 22), 'PRIMER_LEFT_0_TM': 60.94910495105978, 'PRIMER_RIGHT_0_TM': 63.729796086963006, 'PRIMER_LEFT_0_GC_PERCENT': 55.0, 'PRIMER_RIGHT_0_GC_PERCENT': 54.54545454545455, 'PRIMER_LEFT_0_SELF_ANY_TH': 0.0, 'PRIMER_RIGHT_0_SELF_ANY_TH': 23.940233530563603, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 0.0, 'PRIMER_LEFT_0_HAIRPIN_TH': 0.0, 'PRIMER_RIGHT_0_HAIRPIN_TH': 45.488676854316, 'PRIMER_LEFT_0_END_STABILITY': 4.12, 'PRIMER_RIGHT_0_END_STABILITY': 2.92, 'PRIMER_PAIR_0_COMPL_ANY_TH': 23.525035600698857, 'PRIMER_PAIR_0_COMPL_END_TH': 20.267342720304953, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1025}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, high hairpin stability 2, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 27, not in any ok right region 172, ok 209', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 6.198941826581574, 'PRIMER_LEFT_0_PENALTY': 4.6075054315326724, 'PRIMER_RIGHT_0_PENALTY': 1.5914363950489019, 'PRIMER_LEFT_0_PROBLEMS': ' Hairpin stability too high;', 'PRIMER_LEFT_0_SEQUENCE': 'GCTGAGATCGAAAAGCGTGTTGCT', 'PRIMER_RIGHT_0_SEQUENCE': 'TGGCTGCTGGACAATTCGGGA', 'PRIMER_LEFT_0': (0, 24), 'PRIMER_RIGHT_0': (1031, 21), 'PRIMER_LEFT_0_TM': 64.39249456846733, 'PRIMER_RIGHT_0_TM': 64.4085636049511, 'PRIMER_LEFT_0_GC_PERCENT': 50.0, 'PRIMER_RIGHT_0_GC_PERCENT': 57.142857142857146, 'PRIMER_LEFT_0_SELF_ANY_TH': 0.0, 'PRIMER_RIGHT_0_SELF_ANY_TH': 0.0, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 0.0, 'PRIMER_LEFT_0_HAIRPIN_TH': 63.099514384391625, 'PRIMER_RIGHT_0_HAIRPIN_TH': 0.0, 'PRIMER_LEFT_0_END_STABILITY': 3.91, 'PRIMER_RIGHT_0_END_STABILITY': 5.14, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1032}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, low tm 4, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 39, not in any ok right region 172, ok 197', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 12.610259019731643, 'PRIMER_LEFT_0_PENALTY': 10.061720986376315, 'PRIMER_RIGHT_0_PENALTY': 2.548538033355328, 'PRIMER_LEFT_0_SEQUENCE': 'ATGACTAAAACACTTGGTTCCCT', 'PRIMER_RIGHT_0_SEQUENCE': 'AAGCTTGGGCCATTCCTGCT', 'PRIMER_LEFT_0': (0, 23), 'PRIMER_RIGHT_0': (1019, 20), 'PRIMER_LEFT_0_TM': 57.938279013623685, 'PRIMER_RIGHT_0_TM': 62.45146196664467, 'PRIMER_LEFT_0_GC_PERCENT': 39.130434782608695, 'PRIMER_RIGHT_0_GC_PERCENT': 55.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 0.0, 'PRIMER_RIGHT_0_SELF_ANY_TH': 7.26366105581792, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 9.764466054031118, 'PRIMER_LEFT_0_HAIRPIN_TH': 39.28793132834238, 'PRIMER_RIGHT_0_HAIRPIN_TH': 39.730566511905124, 'PRIMER_LEFT_0_END_STABILITY': 4.2, 'PRIMER_RIGHT_0_END_STABILITY': 4.24, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1020}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, low tm 2, high hairpin stability 4, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 25, high hairpin stability 2, not in any ok right region 172, ok 209', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 3.835340222131606, 'PRIMER_LEFT_0_PENALTY': 3.0531575388139345, 'PRIMER_RIGHT_0_PENALTY': 0.7821826833176715, 'PRIMER_LEFT_0_PROBLEMS': ' Hairpin stability too high;', 'PRIMER_LEFT_0_SEQUENCE': 'CTACATGCGATTCTTGGTGCGCG', 'PRIMER_RIGHT_0_SEQUENCE': 'GGTGGTGTCCGAAAACCCGC', 'PRIMER_LEFT_0': (0, 23), 'PRIMER_RIGHT_0': (1023, 20), 'PRIMER_LEFT_0_TM': 64.94684246118607, 'PRIMER_RIGHT_0_TM': 64.21781731668233, 'PRIMER_LEFT_0_GC_PERCENT': 56.52173913043478, 'PRIMER_RIGHT_0_GC_PERCENT': 65.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 0.0, 'PRIMER_RIGHT_0_SELF_ANY_TH': 0.0, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 0.0, 'PRIMER_LEFT_0_HAIRPIN_TH': 51.224837733274626, 'PRIMER_RIGHT_0_HAIRPIN_TH': 35.4429864819399, 'PRIMER_LEFT_0_END_STABILITY': 6.86, 'PRIMER_RIGHT_0_END_STABILITY': 6.13, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1024}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, low tm 4, high hairpin stability 1, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 47, high tm 3, not in any ok right region 172, ok 186', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 9.36879062420013, 'PRIMER_LEFT_0_PENALTY': 9.10487721337637, 'PRIMER_RIGHT_0_PENALTY': 0.2639134108237613, 'PRIMER_LEFT_0_PROBLEMS': ' Hairpin stability too high;', 'PRIMER_LEFT_0_SEQUENCE': 'GCTTTTAAGTTTTCTCGGTGATCCG', 'PRIMER_RIGHT_0_SEQUENCE': 'CGCGCACGCCTCATGGATTC', 'PRIMER_LEFT_0': (0, 25), 'PRIMER_RIGHT_0': (1036, 20), 'PRIMER_LEFT_0_TM': 60.89512278662363, 'PRIMER_RIGHT_0_TM': 64.73608658917624, 'PRIMER_LEFT_0_GC_PERCENT': 44.0, 'PRIMER_RIGHT_0_GC_PERCENT': 65.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 0.0, 'PRIMER_RIGHT_0_SELF_ANY_TH': 8.74883535431286, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 0.0, 'PRIMER_LEFT_0_HAIRPIN_TH': 55.99349097302695, 'PRIMER_RIGHT_0_HAIRPIN_TH': 37.98334977468494, 'PRIMER_LEFT_0_END_STABILITY': 4.18, 'PRIMER_RIGHT_0_END_STABILITY': 2.52, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1037}\n",
      "{}\n",
      "{'PRIMER_LEFT_EXPLAIN': 'considered 8, low tm 1, ok 8', 'PRIMER_RIGHT_EXPLAIN': 'considered 408, low tm 54, not in any ok right region 172, ok 182', 'PRIMER_PAIR_EXPLAIN': 'considered 1, ok 1', 'PRIMER_LEFT_NUM_RETURNED': 1, 'PRIMER_RIGHT_NUM_RETURNED': 1, 'PRIMER_INTERNAL_NUM_RETURNED': 0, 'PRIMER_PAIR_NUM_RETURNED': 1, 'PRIMER_PAIR_0_PENALTY': 10.436286605489158, 'PRIMER_LEFT_0_PENALTY': 5.407095062055419, 'PRIMER_RIGHT_0_PENALTY': 5.02919154343374, 'PRIMER_LEFT_0_SEQUENCE': 'CTACCTATTTGCGCGGTACCACT', 'PRIMER_RIGHT_0_SEQUENCE': 'TCCACATTGTTCGTGATCGCTTCGA', 'PRIMER_LEFT_0': (0, 23), 'PRIMER_RIGHT_0': (1034, 25), 'PRIMER_LEFT_0_TM': 62.59290493794458, 'PRIMER_RIGHT_0_TM': 64.97080845656626, 'PRIMER_LEFT_0_GC_PERCENT': 52.17391304347826, 'PRIMER_RIGHT_0_GC_PERCENT': 48.0, 'PRIMER_LEFT_0_SELF_ANY_TH': 13.539476625797363, 'PRIMER_RIGHT_0_SELF_ANY_TH': 4.4829346765852165, 'PRIMER_LEFT_0_SELF_END_TH': 0.0, 'PRIMER_RIGHT_0_SELF_END_TH': 1.6655968014014206, 'PRIMER_LEFT_0_HAIRPIN_TH': 32.356514321702605, 'PRIMER_RIGHT_0_HAIRPIN_TH': 41.134684649373696, 'PRIMER_LEFT_0_END_STABILITY': 4.0, 'PRIMER_RIGHT_0_END_STABILITY': 3.71, 'PRIMER_PAIR_0_COMPL_ANY_TH': 0.0, 'PRIMER_PAIR_0_COMPL_END_TH': 0.0, 'PRIMER_PAIR_0_PRODUCT_SIZE': 1035}\n",
      "{}\n",
      "Index(['primer_f_seq_(5'-3')', 'primer_f_Tm', 'primer_r_seq_(5'-3')',\n",
      "       'primer_r_Tm', 'product_size', 'product_value', 'Name', 'Region',\n",
      "       'Type'],\n",
      "      dtype='object')\n",
      "--1.执行单质粒系统,--2.执行双质粒系统,---0.执行单、双质粒系统都执行------现在正在执行的情况： 0\n"
     ]
    }
   ],
   "source": [
    "def work(data):\n",
    "    chopchop_input = data['chopchop_input']\n",
    "    print(os.getcwd())\n",
    "    \n",
    "    #uha_dha参数\n",
    "    uha_dha_params = data['uha_dha_config']\n",
    "    \n",
    "    # enzyme_path = parent_base_path +'/'+ data['enzyme_path']\n",
    "    sgRNA_result_path = data['sgRNA_result_path']\n",
    "    plasmid_file_1 = data['one_plasmid_file_path']\n",
    "    plasmid_file_2 = data['no_ccdb_plasmid']\n",
    "    plasmid_file_3 = data['no_sgRNA_plasmid']\n",
    "    \n",
    "    #plasmid label\n",
    "    ccdb_label = data['plasmid_label']['ccdb_label']\n",
    "    promoter_terminator_label = data['plasmid_label']['promoter_terminator_label']\n",
    "    n_20_label = data['plasmid_label']['n_20_label'] \n",
    "    \n",
    "    #检查质粒标签\n",
    "    if ccdb_label == promoter_terminator_label or ccdb_label == n_20_label:\n",
    "        errorMessage = 'There are duplicates in the three plasmid label you entered'\n",
    "        raise ValueError(errorMessage)\n",
    "    elif promoter_terminator_label == n_20_label or promoter_terminator_label == ccdb_label:\n",
    "        errorMessage = 'There are duplicates in the three plasmid label you entered'\n",
    "        raise ValueError(errorMessage)\n",
    "    \n",
    "    \n",
    "    one_plasmid_file_path=''\n",
    "    no_ccdb_plasmid=''\n",
    "    no_sgRNA_plasmid=''\n",
    "    \n",
    "    #0.检查质粒,自动判断质粒类型\n",
    "    if plasmid_file_1 != '':\n",
    "        plasmid_type = check_plasmid(plasmid_file_1, ccdb_label, promoter_terminator_label, n_20_label)\n",
    "        if plasmid_type == 'one_plasmid_file_path':\n",
    "            one_plasmid_file_path = plasmid_file_1\n",
    "        elif plasmid_type == 'no_ccdb_plasmid':\n",
    "            no_ccdb_plasmid = plasmid_file_1\n",
    "        elif plasmid_type == 'no_sgRNA_plasmid':\n",
    "            no_sgRNA_plasmid = plasmid_file_1\n",
    "    \n",
    "    \n",
    "    if plasmid_file_2 != '':\n",
    "        plasmid_type = check_plasmid(plasmid_file_2, ccdb_label, promoter_terminator_label, n_20_label)\n",
    "        if plasmid_type == 'one_plasmid_file_path':\n",
    "            one_plasmid_file_path = plasmid_file_2\n",
    "        elif plasmid_type == 'no_ccdb_plasmid':\n",
    "            no_ccdb_plasmid = plasmid_file_2\n",
    "        elif plasmid_type == 'no_sgRNA_plasmid':\n",
    "            no_sgRNA_plasmid = plasmid_file_2\n",
    "        \n",
    "\n",
    "    if plasmid_file_3 != '':\n",
    "        plasmid_type = check_plasmid(plasmid_file_3, ccdb_label, promoter_terminator_label, n_20_label)    \n",
    "        if plasmid_type == 'one_plasmid_file_path':\n",
    "            one_plasmid_file_path = plasmid_file_3\n",
    "        elif plasmid_type == 'no_ccdb_plasmid':\n",
    "            no_ccdb_plasmid = plasmid_file_3\n",
    "        elif plasmid_type == 'no_sgRNA_plasmid':\n",
    "            no_sgRNA_plasmid = plasmid_file_3\n",
    "    \n",
    "\n",
    "    #1.获取引物，检查引物\n",
    "    sgRNA_primer_json = data['sgRNA_primer_json']\n",
    "    ccdb_primer_json = data['ccdb_primer_json']  \n",
    "    primer_json = data['primer_json']  \n",
    "    \n",
    "    if primer_json != {} and one_plasmid_file_path !='':\n",
    "        #取质粒骨架，质粒label：ccdb_label、promoter_terminator_label、n_20_label\n",
    "        plasmid_backbone =  p_d_seq.get_plasmid_backbone_by_labels(one_plasmid_file_path, ccdb_label=ccdb_label, promoter_terminator_label=promoter_terminator_label, n_20_label=n_20_label)\n",
    "        failture_seq_json = check_quality_control(plasmid_backbone, primer_json)\n",
    "        if failture_seq_json != {} or failture_seq_json == None:\n",
    "            raise ValueError(failture_seq_json)\n",
    "    \n",
    "    if sgRNA_primer_json !={} and no_ccdb_plasmid !='':\n",
    "        plasmid_backbone =  p_d_seq.get_plasmid_backbone_by_labels(no_ccdb_plasmid, ccdb_label=ccdb_label, promoter_terminator_label=promoter_terminator_label, n_20_label=n_20_label)\n",
    "        failture_seq_json = check_quality_control(plasmid_backbone, sgRNA_primer_json)\n",
    "        if failture_seq_json != {}:\n",
    "            raise ValueError(failture_seq_json)\n",
    "            \n",
    "    if ccdb_primer_json !={} and no_sgRNA_plasmid != '':\n",
    "        plasmid_backbone =  p_d_seq.get_plasmid_backbone_by_labels(no_sgRNA_plasmid, ccdb_label=ccdb_label, promoter_terminator_label=promoter_terminator_label, n_20_label=n_20_label)\n",
    "        failture_seq_json = check_quality_control(plasmid_backbone, ccdb_primer_json)\n",
    "        if failture_seq_json != {}:\n",
    "            raise ValueError(failture_seq_json)\n",
    "\n",
    "\n",
    "    #2.根据获取的label，取出region序列，且检查区域序列\n",
    "    region_label = data['region_label']\n",
    "    sgRNA_region_label = data['sgRNA_region_label']\n",
    "    ccdb_region_label = data['ccdb_region_label']\n",
    "\n",
    "    region_seq_json = {}\n",
    "    sgRNA_region_seq_json = {}\n",
    "    ccdb_region_seq_json = {}\n",
    "    \n",
    "    if region_label !=\"\" and one_plasmid_file_path !=\"\":\n",
    "        seq = su.get_sequence_by_feature_label(one_plasmid_file_path, region_label)\n",
    "        region_seq_json = {'region1':seq}\n",
    "        #检查\n",
    "        failture_seq_json = check_quality_control(plasmid_backbone, region_seq_json)\n",
    "        if failture_seq_json != {}:\n",
    "            raise ValueError(failture_seq_json)\n",
    "    \n",
    "    if sgRNA_region_label !=\"\" and no_ccdb_plasmid !=\"\":\n",
    "        seq = su.get_sequence_by_feature_label(no_ccdb_plasmid, sgRNA_region_label)\n",
    "        sgRNA_region_seq_json = {'region1':seq}\n",
    "        #检查\n",
    "        failture_seq_json = check_quality_control(plasmid_backbone, sgRNA_region_seq_json)\n",
    "        if failture_seq_json != {}:\n",
    "            raise ValueError(failture_seq_json)\n",
    "    \n",
    "    if ccdb_region_label !=\"\" and no_sgRNA_plasmid !=\"\":\n",
    "        seq = su.get_sequence_by_feature_label(no_sgRNA_plasmid, ccdb_region_label)\n",
    "        ccdb_region_seq_json = {'region1':seq}\n",
    "        #检查\n",
    "        failture_seq_json = check_quality_control(plasmid_backbone, ccdb_region_seq_json)\n",
    "        if failture_seq_json != {}:\n",
    "            raise ValueError(failture_seq_json)\n",
    "\n",
    "\n",
    "\n",
    "    #genome\n",
    "    genome_path = data['ref_genome'] \n",
    "\n",
    "    #配置引物参数\n",
    "    # config.S_GLOBAL_ARGS = data['S_GLOBAL_ARGS']\n",
    "    # config.Q_ARGS = data['Q_ARGS']\n",
    "    config.UHA_ARGS = data['UHA_ARGS']\n",
    "    config.SEQ_ALTERED_ARGS = data['SEQ_ALTERED_ARGS']\n",
    "    config.DHA_ARGS = data['DHA_ARGS']\n",
    "    \n",
    "    #4.判断使用何种技术方法：PCR、OLIGO\n",
    "    if data.get('UP_SGRNA_ARGS').get('PRIMER_OPT_TM') == '' or data.get('UP_SGRNA_ARGS') == {} or data.get('DOWN_SGRNA_ARGS') == {}:\n",
    "        method = 'OLIGO'\n",
    "        print('执行方法：', method)\n",
    "    else:\n",
    "        method = 'PCR'\n",
    "        print('执行方法：', method)\n",
    "        config.UP_SGRNA_ARGS = data['UP_SGRNA_ARGS']\n",
    "        config.DOWN_SGRNA_ARGS = data['DOWN_SGRNA_ARGS']\n",
    "\n",
    "    config.PLASMID_Q_ARGS = data['PLASMID_Q_ARGS']\n",
    "    config.GENOME_Q_ARGS = data['GENOME_Q_ARGS']   \n",
    "    \n",
    "    #配置输出参数\n",
    "    output = data['edit_sequence_design_workdir']\n",
    "    if not os.path.exists(output):\n",
    "        os.makedirs(output)\n",
    "    scene = data['scene']  \n",
    "\n",
    "  \n",
    "    # 5.read 编辑序列信息,给chopchop输入加uha、dha信息\n",
    "    info_input_df = read_chopchopInput_add_uha_dha(genome_path, chopchop_input, uha_dha_params)\n",
    "    \n",
    "    base_path = '/home/yanghe/program/edit_sequence_design'\n",
    "    enzyme_path =  base_path + '/input/enzyme.csv'      \n",
    "    enzyme_df = su.del_Unnamed(pd.read_csv(enzyme_path))\n",
    "    enzyme = data['enzyme']\n",
    "    enzyme_name = enzyme['enzyme_name']     \n",
    "    #6.检查酶的相关信息\n",
    "    enzyme_df = check_enzyme(enzyme, enzyme_df)\n",
    "\n",
    "\n",
    "    # 7.判断不同的场景，提取用户选择的sgRNA     \n",
    "    if scene == 'only_primer': \n",
    "        sgRNA = info_input_df[['name','crrna','region']].rename(columns={'crrna':\"Target sequence\",\"name\":\"Name\",\"region\":\"Region\"})\n",
    "        sgRNA['Rev Target sequence'] = sgRNA['Target sequence'].apply(lambda x: su.revComp(x)) \n",
    "\n",
    "    elif scene == 'both_sgRNA_primer':\n",
    "        selected_sgRNA_result = data['sgRNA_result']\n",
    "        sgRNA = p_d_seq.extract_sgRNA_from_chopchop(sgRNA_result_path, selected_sgRNA_result)\n",
    "    \n",
    "\n",
    "    # 8.设计源生同源臂引物\n",
    "    uha_dha_primer_df = extract_uha_dha_primer(info_input_df, sgRNA)   \n",
    "\n",
    "    # 9.提取同源臂\n",
    "    uha_dha_info_primer_df, uha_dha_df, uha_dha_sgRNA_df, info_df = extract_uha_dha(info_input_df, uha_dha_primer_df, sgRNA)\n",
    "  \n",
    "    #10.判断质粒的执行类型    \n",
    "    if  one_plasmid_file_path != '' and no_ccdb_plasmid == '' and no_sgRNA_plasmid == '':\n",
    "        plasmid_system_type = 1\n",
    "    elif  one_plasmid_file_path =='' and no_ccdb_plasmid != '' and no_sgRNA_plasmid != '':\n",
    "        plasmid_system_type = 2\n",
    "    elif  one_plasmid_file_path !='' and no_ccdb_plasmid !='' and no_sgRNA_plasmid !='':\n",
    "        plasmid_system_type = 0\n",
    "  \n",
    "    print('--1.执行单质粒系统,--2.执行双质粒系统,---0.执行单、双质粒系统都执行------现在正在执行的情况：',plasmid_system_type) \n",
    "\n",
    "    \n",
    "    plasmid_system_type = 1\n",
    "    if plasmid_system_type == 1:\n",
    "\n",
    "        #质粒引物的设计类型：1---用户指定范围，2----无需用户指定范围，3----用户指定额外引物\n",
    "        if region_seq_json == {} and  primer_json == {}:\n",
    "            plasmid_primer_desgin_type = 2\n",
    "        elif region_seq_json != {} and primer_json == {}:\n",
    "            plasmid_primer_desgin_type = 1\n",
    "        elif primer_json != {} and region_seq_json == {}:\n",
    "            plasmid_primer_desgin_type = 3\n",
    "        else:\n",
    "            plasmid_primer_desgin_type = 1\n",
    "\n",
    "        # 6.执行单质粒系统\n",
    "        one_plasmid_output_path = execute_one_plasmid_system(   \n",
    "                                                                plasmid_primer_desgin_type,\n",
    "                                                                region_seq_json,\n",
    "                                                                primer_json,\n",
    "                                                                one_plasmid_file_path,\n",
    "                                                                info_df,\n",
    "                                                                info_input_df,\n",
    "                                                                uha_dha_df,\n",
    "                                                                uha_dha_sgRNA_df,\n",
    "                                                                uha_dha_info_primer_df,\n",
    "                                                                uha_dha_primer_df,\n",
    "                                                                enzyme_df,\n",
    "                                                                enzyme_name,\n",
    "                                                                output\n",
    "                                                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bfce5f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60e40487",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed72d16b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
